<!DOCTYPE html>

<html>

<head>

<meta charset="utf-8" />
<meta name="generator" content="pandoc" />
<meta http-equiv="X-UA-Compatible" content="IE=EDGE" />

<meta name="viewport" content="width=device-width, initial-scale=1" />


<meta name="date" content="2024-09-20" />

<title>Capture all matches in a single subject string</title>

<script>// Pandoc 2.9 adds attributes on both header and div. We remove the former (to
// be compatible with the behavior of Pandoc < 2.8).
document.addEventListener('DOMContentLoaded', function(e) {
  var hs = document.querySelectorAll("div.section[class*='level'] > :first-child");
  var i, h, a;
  for (i = 0; i < hs.length; i++) {
    h = hs[i];
    if (!/^h[1-6]$/i.test(h.tagName)) continue;  // it should be a header h1-h6
    a = h.attributes;
    while (a.length > 0) h.removeAttribute(a[0].name);
  }
});
</script>

<style type="text/css">
code{white-space: pre-wrap;}
span.smallcaps{font-variant: small-caps;}
span.underline{text-decoration: underline;}
div.column{display: inline-block; vertical-align: top; width: 50%;}
div.hanging-indent{margin-left: 1.5em; text-indent: -1.5em;}
ul.task-list{list-style: none;}
</style>



<style type="text/css">
code {
white-space: pre;
}
.sourceCode {
overflow: visible;
}
</style>
<style type="text/css" data-origin="pandoc">
pre > code.sourceCode { white-space: pre; position: relative; }
pre > code.sourceCode > span { line-height: 1.25; }
pre > code.sourceCode > span:empty { height: 1.2em; }
.sourceCode { overflow: visible; }
code.sourceCode > span { color: inherit; text-decoration: inherit; }
div.sourceCode { margin: 1em 0; }
pre.sourceCode { margin: 0; }
@media screen {
div.sourceCode { overflow: auto; }
}
@media print {
pre > code.sourceCode { white-space: pre-wrap; }
pre > code.sourceCode > span { display: inline-block; text-indent: -5em; padding-left: 5em; }
}
pre.numberSource code
{ counter-reset: source-line 0; }
pre.numberSource code > span
{ position: relative; left: -4em; counter-increment: source-line; }
pre.numberSource code > span > a:first-child::before
{ content: counter(source-line);
position: relative; left: -1em; text-align: right; vertical-align: baseline;
border: none; display: inline-block;
-webkit-touch-callout: none; -webkit-user-select: none;
-khtml-user-select: none; -moz-user-select: none;
-ms-user-select: none; user-select: none;
padding: 0 4px; width: 4em;
color: #aaaaaa;
}
pre.numberSource { margin-left: 3em; border-left: 1px solid #aaaaaa; padding-left: 4px; }
div.sourceCode
{ }
@media screen {
pre > code.sourceCode > span > a:first-child::before { text-decoration: underline; }
}
code span.al { color: #ff0000; font-weight: bold; } 
code span.an { color: #60a0b0; font-weight: bold; font-style: italic; } 
code span.at { color: #7d9029; } 
code span.bn { color: #40a070; } 
code span.bu { color: #008000; } 
code span.cf { color: #007020; font-weight: bold; } 
code span.ch { color: #4070a0; } 
code span.cn { color: #880000; } 
code span.co { color: #60a0b0; font-style: italic; } 
code span.cv { color: #60a0b0; font-weight: bold; font-style: italic; } 
code span.do { color: #ba2121; font-style: italic; } 
code span.dt { color: #902000; } 
code span.dv { color: #40a070; } 
code span.er { color: #ff0000; font-weight: bold; } 
code span.ex { } 
code span.fl { color: #40a070; } 
code span.fu { color: #06287e; } 
code span.im { color: #008000; font-weight: bold; } 
code span.in { color: #60a0b0; font-weight: bold; font-style: italic; } 
code span.kw { color: #007020; font-weight: bold; } 
code span.op { color: #666666; } 
code span.ot { color: #007020; } 
code span.pp { color: #bc7a00; } 
code span.sc { color: #4070a0; } 
code span.ss { color: #bb6688; } 
code span.st { color: #4070a0; } 
code span.va { color: #19177c; } 
code span.vs { color: #4070a0; } 
code span.wa { color: #60a0b0; font-weight: bold; font-style: italic; } 
</style>
<script>
// apply pandoc div.sourceCode style to pre.sourceCode instead
(function() {
  var sheets = document.styleSheets;
  for (var i = 0; i < sheets.length; i++) {
    if (sheets[i].ownerNode.dataset["origin"] !== "pandoc") continue;
    try { var rules = sheets[i].cssRules; } catch (e) { continue; }
    var j = 0;
    while (j < rules.length) {
      var rule = rules[j];
      // check if there is a div.sourceCode rule
      if (rule.type !== rule.STYLE_RULE || rule.selectorText !== "div.sourceCode") {
        j++;
        continue;
      }
      var style = rule.style.cssText;
      // check if color or background-color is set
      if (rule.style.color === '' && rule.style.backgroundColor === '') {
        j++;
        continue;
      }
      // replace div.sourceCode by a pre.sourceCode rule
      sheets[i].deleteRule(j);
      sheets[i].insertRule('pre.sourceCode{' + style + '}', j);
    }
  }
})();
</script>




<style type="text/css">body {
background-color: #fff;
margin: 1em auto;
max-width: 700px;
overflow: visible;
padding-left: 2em;
padding-right: 2em;
font-family: "Open Sans", "Helvetica Neue", Helvetica, Arial, sans-serif;
font-size: 14px;
line-height: 1.35;
}
#TOC {
clear: both;
margin: 0 0 10px 10px;
padding: 4px;
width: 400px;
border: 1px solid #CCCCCC;
border-radius: 5px;
background-color: #f6f6f6;
font-size: 13px;
line-height: 1.3;
}
#TOC .toctitle {
font-weight: bold;
font-size: 15px;
margin-left: 5px;
}
#TOC ul {
padding-left: 40px;
margin-left: -1.5em;
margin-top: 5px;
margin-bottom: 5px;
}
#TOC ul ul {
margin-left: -2em;
}
#TOC li {
line-height: 16px;
}
table {
margin: 1em auto;
border-width: 1px;
border-color: #DDDDDD;
border-style: outset;
border-collapse: collapse;
}
table th {
border-width: 2px;
padding: 5px;
border-style: inset;
}
table td {
border-width: 1px;
border-style: inset;
line-height: 18px;
padding: 5px 5px;
}
table, table th, table td {
border-left-style: none;
border-right-style: none;
}
table thead, table tr.even {
background-color: #f7f7f7;
}
p {
margin: 0.5em 0;
}
blockquote {
background-color: #f6f6f6;
padding: 0.25em 0.75em;
}
hr {
border-style: solid;
border: none;
border-top: 1px solid #777;
margin: 28px 0;
}
dl {
margin-left: 0;
}
dl dd {
margin-bottom: 13px;
margin-left: 13px;
}
dl dt {
font-weight: bold;
}
ul {
margin-top: 0;
}
ul li {
list-style: circle outside;
}
ul ul {
margin-bottom: 0;
}
pre, code {
background-color: #f7f7f7;
border-radius: 3px;
color: #333;
white-space: pre-wrap; 
}
pre {
border-radius: 3px;
margin: 5px 0px 10px 0px;
padding: 10px;
}
pre:not([class]) {
background-color: #f7f7f7;
}
code {
font-family: Consolas, Monaco, 'Courier New', monospace;
font-size: 85%;
}
p > code, li > code {
padding: 2px 0px;
}
div.figure {
text-align: center;
}
img {
background-color: #FFFFFF;
padding: 2px;
border: 1px solid #DDDDDD;
border-radius: 3px;
border: 1px solid #CCCCCC;
margin: 0 5px;
}
h1 {
margin-top: 0;
font-size: 35px;
line-height: 40px;
}
h2 {
border-bottom: 4px solid #f7f7f7;
padding-top: 10px;
padding-bottom: 2px;
font-size: 145%;
}
h3 {
border-bottom: 2px solid #f7f7f7;
padding-top: 10px;
font-size: 120%;
}
h4 {
border-bottom: 1px solid #f7f7f7;
margin-left: 8px;
font-size: 105%;
}
h5, h6 {
border-bottom: 1px solid #ccc;
font-size: 105%;
}
a {
color: #0033dd;
text-decoration: none;
}
a:hover {
color: #6666ff; }
a:visited {
color: #800080; }
a:visited:hover {
color: #BB00BB; }
a[href^="http:"] {
text-decoration: underline; }
a[href^="https:"] {
text-decoration: underline; }

code > span.kw { color: #555; font-weight: bold; } 
code > span.dt { color: #902000; } 
code > span.dv { color: #40a070; } 
code > span.bn { color: #d14; } 
code > span.fl { color: #d14; } 
code > span.ch { color: #d14; } 
code > span.st { color: #d14; } 
code > span.co { color: #888888; font-style: italic; } 
code > span.ot { color: #007020; } 
code > span.al { color: #ff0000; font-weight: bold; } 
code > span.fu { color: #900; font-weight: bold; } 
code > span.er { color: #a61717; background-color: #e3d2d2; } 
</style>




</head>

<body>




<h1 class="title toc-ignore">Capture all matches in a single subject
string</h1>
<h4 class="date">2024-09-20</h4>



<p>The <code>nc::capture_all_str</code> function is for the common case
of extracting each match from a multi-line text file (a single large
subject string). In this section we demonstrate how to extract data
tables from such loosely structured text data. For example we consider
the following <a href="http://genome.cse.ucsc.edu/goldenPath/help/hgTrackHubHelp.html">track
hub</a> meta-data file:</p>
<div class="sourceCode" id="cb1"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb1-1"><a href="#cb1-1" tabindex="-1"></a>trackDb.txt.gz <span class="ot">&lt;-</span> <span class="fu">system.file</span>(</span>
<span id="cb1-2"><a href="#cb1-2" tabindex="-1"></a>  <span class="st">&quot;extdata&quot;</span>, <span class="st">&quot;trackDb.txt.gz&quot;</span>, <span class="at">package=</span><span class="st">&quot;nc&quot;</span>)</span>
<span id="cb1-3"><a href="#cb1-3" tabindex="-1"></a>trackDb.vec <span class="ot">&lt;-</span> <span class="fu">readLines</span>(trackDb.txt.gz)</span></code></pre></div>
<p>Some representative lines from that file are shown below.</p>
<div class="sourceCode" id="cb2"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb2-1"><a href="#cb2-1" tabindex="-1"></a><span class="fu">cat</span>(trackDb.vec[<span class="dv">78</span><span class="sc">:</span><span class="dv">107</span>], <span class="at">sep=</span><span class="st">&quot;</span><span class="sc">\n</span><span class="st">&quot;</span>)</span>
<span id="cb2-2"><a href="#cb2-2" tabindex="-1"></a><span class="co">#&gt; track peaks_summary</span></span>
<span id="cb2-3"><a href="#cb2-3" tabindex="-1"></a><span class="co">#&gt; type bigBed 5</span></span>
<span id="cb2-4"><a href="#cb2-4" tabindex="-1"></a><span class="co">#&gt; shortLabel _model_peaks_summary</span></span>
<span id="cb2-5"><a href="#cb2-5" tabindex="-1"></a><span class="co">#&gt; longLabel Regions with a peak in at least one sample</span></span>
<span id="cb2-6"><a href="#cb2-6" tabindex="-1"></a><span class="co">#&gt; visibility pack</span></span>
<span id="cb2-7"><a href="#cb2-7" tabindex="-1"></a><span class="co">#&gt; itemRgb off</span></span>
<span id="cb2-8"><a href="#cb2-8" tabindex="-1"></a><span class="co">#&gt; spectrum on</span></span>
<span id="cb2-9"><a href="#cb2-9" tabindex="-1"></a><span class="co">#&gt; bigDataUrl http://hubs.hpc.mcgill.ca/~thocking/PeakSegFPOP-/peaks_summary.bigBed</span></span>
<span id="cb2-10"><a href="#cb2-10" tabindex="-1"></a><span class="co">#&gt; </span></span>
<span id="cb2-11"><a href="#cb2-11" tabindex="-1"></a><span class="co">#&gt; </span></span>
<span id="cb2-12"><a href="#cb2-12" tabindex="-1"></a><span class="co">#&gt;  track bcell_McGill0091</span></span>
<span id="cb2-13"><a href="#cb2-13" tabindex="-1"></a><span class="co">#&gt;  parent bcell</span></span>
<span id="cb2-14"><a href="#cb2-14" tabindex="-1"></a><span class="co">#&gt;  container multiWig</span></span>
<span id="cb2-15"><a href="#cb2-15" tabindex="-1"></a><span class="co">#&gt;  type bigWig</span></span>
<span id="cb2-16"><a href="#cb2-16" tabindex="-1"></a><span class="co">#&gt;  shortLabel bcell_McGill0091</span></span>
<span id="cb2-17"><a href="#cb2-17" tabindex="-1"></a><span class="co">#&gt;  longLabel bcell | McGill0091</span></span>
<span id="cb2-18"><a href="#cb2-18" tabindex="-1"></a><span class="co">#&gt;  graphType points</span></span>
<span id="cb2-19"><a href="#cb2-19" tabindex="-1"></a><span class="co">#&gt;  aggregate transparentOverlay</span></span>
<span id="cb2-20"><a href="#cb2-20" tabindex="-1"></a><span class="co">#&gt;  showSubtrackColorOnUi on</span></span>
<span id="cb2-21"><a href="#cb2-21" tabindex="-1"></a><span class="co">#&gt;  maxHeightPixels 25:12:8</span></span>
<span id="cb2-22"><a href="#cb2-22" tabindex="-1"></a><span class="co">#&gt;  visibility full</span></span>
<span id="cb2-23"><a href="#cb2-23" tabindex="-1"></a><span class="co">#&gt;  autoScale on</span></span>
<span id="cb2-24"><a href="#cb2-24" tabindex="-1"></a><span class="co">#&gt; </span></span>
<span id="cb2-25"><a href="#cb2-25" tabindex="-1"></a><span class="co">#&gt;   track bcell_McGill0091Coverage</span></span>
<span id="cb2-26"><a href="#cb2-26" tabindex="-1"></a><span class="co">#&gt;   bigDataUrl http://hubs.hpc.mcgill.ca/~thocking/PeakSegFPOP-/samples/bcell/McGill0091/coverage.bigWig</span></span>
<span id="cb2-27"><a href="#cb2-27" tabindex="-1"></a><span class="co">#&gt;   shortLabel bcell_McGill0091Coverage</span></span>
<span id="cb2-28"><a href="#cb2-28" tabindex="-1"></a><span class="co">#&gt;   longLabel bcell | McGill0091 | Coverage</span></span>
<span id="cb2-29"><a href="#cb2-29" tabindex="-1"></a><span class="co">#&gt;   parent bcell_McGill0091</span></span>
<span id="cb2-30"><a href="#cb2-30" tabindex="-1"></a><span class="co">#&gt;   type bigWig</span></span>
<span id="cb2-31"><a href="#cb2-31" tabindex="-1"></a><span class="co">#&gt;   color 141,211,199</span></span></code></pre></div>
<div id="match-all-tracks-in-the-text-file" class="section level2">
<h2>Match all tracks in the text file</h2>
<p>Each block of text begins with “track” and includes several lines of
data before the block ends with two consecutive newlines. That pattern
is coded below using a regex:</p>
<div class="sourceCode" id="cb3"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb3-1"><a href="#cb3-1" tabindex="-1"></a>tracks.dt <span class="ot">&lt;-</span> nc<span class="sc">::</span><span class="fu">capture_all_str</span>(</span>
<span id="cb3-2"><a href="#cb3-2" tabindex="-1"></a>  trackDb.vec, </span>
<span id="cb3-3"><a href="#cb3-3" tabindex="-1"></a>  <span class="st">&quot;track &quot;</span>,</span>
<span id="cb3-4"><a href="#cb3-4" tabindex="-1"></a>  <span class="at">track=</span><span class="st">&quot;</span><span class="sc">\\</span><span class="st">S+&quot;</span>,</span>
<span id="cb3-5"><a href="#cb3-5" tabindex="-1"></a>  <span class="at">fields=</span><span class="st">&quot;(?:</span><span class="sc">\n</span><span class="st">[^</span><span class="sc">\n</span><span class="st">]+)*&quot;</span>,</span>
<span id="cb3-6"><a href="#cb3-6" tabindex="-1"></a>  <span class="st">&quot;</span><span class="sc">\n</span><span class="st">&quot;</span>)</span>
<span id="cb3-7"><a href="#cb3-7" tabindex="-1"></a><span class="fu">str</span>(tracks.dt)</span>
<span id="cb3-8"><a href="#cb3-8" tabindex="-1"></a><span class="co">#&gt; Classes &#39;data.table&#39; and &#39;data.frame&#39;:   123 obs. of  2 variables:</span></span>
<span id="cb3-9"><a href="#cb3-9" tabindex="-1"></a><span class="co">#&gt;  $ track : chr  &quot;bcell&quot; &quot;kidneyCancer&quot; &quot;kidney&quot; &quot;leukemiaCD19CD10BCells&quot; ...</span></span>
<span id="cb3-10"><a href="#cb3-10" tabindex="-1"></a><span class="co">#&gt;  $ fields: chr  &quot;\nsuperTrack on show\nshortLabel bcell\nlongLabel bcell ChIP-seq samples&quot; &quot;\nsuperTrack on show\nshortLabel kidneyCancer\nlongLabel kidneyCancer ChIP-seq samples&quot; &quot;\nsuperTrack on show\nshortLabel kidney\nlongLabel kidney ChIP-seq samples&quot; &quot;\nsuperTrack on show\nshortLabel leukemiaCD19CD10BCells\nlongLabel leukemiaCD19CD10BCells ChIP-seq samples&quot; ...</span></span>
<span id="cb3-11"><a href="#cb3-11" tabindex="-1"></a><span class="co">#&gt;  - attr(*, &quot;.internal.selfref&quot;)=&lt;externalptr&gt;</span></span></code></pre></div>
<p>The result is a data.table with one row for each track block that
matches the regex. There are two character columns: <code>track</code>
is a unique name, and <code>fields</code> is a string with the rest of
the data in that block:</p>
<div class="sourceCode" id="cb4"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb4-1"><a href="#cb4-1" tabindex="-1"></a>tracks.dt[, .(track, <span class="at">fields.start=</span><span class="fu">substr</span>(fields, <span class="dv">1</span>, <span class="dv">30</span>))]</span>
<span id="cb4-2"><a href="#cb4-2" tabindex="-1"></a><span class="co">#&gt;                         track                     fields.start</span></span>
<span id="cb4-3"><a href="#cb4-3" tabindex="-1"></a><span class="co">#&gt;                        &lt;char&gt;                           &lt;char&gt;</span></span>
<span id="cb4-4"><a href="#cb4-4" tabindex="-1"></a><span class="co">#&gt;   1:                    bcell \nsuperTrack on show\nshortLabel</span></span>
<span id="cb4-5"><a href="#cb4-5" tabindex="-1"></a><span class="co">#&gt;   2:             kidneyCancer \nsuperTrack on show\nshortLabel</span></span>
<span id="cb4-6"><a href="#cb4-6" tabindex="-1"></a><span class="co">#&gt;   3:                   kidney \nsuperTrack on show\nshortLabel</span></span>
<span id="cb4-7"><a href="#cb4-7" tabindex="-1"></a><span class="co">#&gt;   4:   leukemiaCD19CD10BCells \nsuperTrack on show\nshortLabel</span></span>
<span id="cb4-8"><a href="#cb4-8" tabindex="-1"></a><span class="co">#&gt;   5:                 monocyte \nsuperTrack on show\nshortLabel</span></span>
<span id="cb4-9"><a href="#cb4-9" tabindex="-1"></a><span class="co">#&gt;  ---                                                          </span></span>
<span id="cb4-10"><a href="#cb4-10" tabindex="-1"></a><span class="co">#&gt; 119: tcell_McGill0106Coverage  \n  bigDataUrl http://hubs.hpc.</span></span>
<span id="cb4-11"><a href="#cb4-11" tabindex="-1"></a><span class="co">#&gt; 120:    tcell_McGill0106Peaks  \n  bigDataUrl http://hubs.hpc.</span></span>
<span id="cb4-12"><a href="#cb4-12" tabindex="-1"></a><span class="co">#&gt; 121:         tcell_McGill0107 \n parent tcell\n container mult</span></span>
<span id="cb4-13"><a href="#cb4-13" tabindex="-1"></a><span class="co">#&gt; 122: tcell_McGill0107Coverage  \n  bigDataUrl http://hubs.hpc.</span></span>
<span id="cb4-14"><a href="#cb4-14" tabindex="-1"></a><span class="co">#&gt; 123:    tcell_McGill0107Peaks  \n  bigDataUrl http://hubs.hpc.</span></span></code></pre></div>
</div>
<div id="match-all-fields-in-each-track" class="section level2">
<h2>Match all fields in each track</h2>
<p>Each block has a variable number of lines/fields. Each line starts
with a field name, followed by a space, followed by the field value.
That regex is coded below:</p>
<div class="sourceCode" id="cb5"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb5-1"><a href="#cb5-1" tabindex="-1"></a>(fields.dt <span class="ot">&lt;-</span> tracks.dt[, nc<span class="sc">::</span><span class="fu">capture_all_str</span>(</span>
<span id="cb5-2"><a href="#cb5-2" tabindex="-1"></a>  fields,</span>
<span id="cb5-3"><a href="#cb5-3" tabindex="-1"></a>  <span class="st">&quot;</span><span class="sc">\\</span><span class="st">s+&quot;</span>,</span>
<span id="cb5-4"><a href="#cb5-4" tabindex="-1"></a>  <span class="at">variable=</span><span class="st">&quot;.*?&quot;</span>,</span>
<span id="cb5-5"><a href="#cb5-5" tabindex="-1"></a>  <span class="st">&quot; &quot;</span>,</span>
<span id="cb5-6"><a href="#cb5-6" tabindex="-1"></a>  <span class="at">value=</span><span class="st">&quot;[^</span><span class="sc">\n</span><span class="st">]+&quot;</span>),  </span>
<span id="cb5-7"><a href="#cb5-7" tabindex="-1"></a>  <span class="at">by=</span>track])</span>
<span id="cb5-8"><a href="#cb5-8" tabindex="-1"></a><span class="co">#&gt;                      track   variable                      value</span></span>
<span id="cb5-9"><a href="#cb5-9" tabindex="-1"></a><span class="co">#&gt;                     &lt;char&gt;     &lt;char&gt;                     &lt;char&gt;</span></span>
<span id="cb5-10"><a href="#cb5-10" tabindex="-1"></a><span class="co">#&gt;   1:                 bcell superTrack                    on show</span></span>
<span id="cb5-11"><a href="#cb5-11" tabindex="-1"></a><span class="co">#&gt;   2:                 bcell shortLabel                      bcell</span></span>
<span id="cb5-12"><a href="#cb5-12" tabindex="-1"></a><span class="co">#&gt;   3:                 bcell  longLabel     bcell ChIP-seq samples</span></span>
<span id="cb5-13"><a href="#cb5-13" tabindex="-1"></a><span class="co">#&gt;   4:          kidneyCancer superTrack                    on show</span></span>
<span id="cb5-14"><a href="#cb5-14" tabindex="-1"></a><span class="co">#&gt;   5:          kidneyCancer shortLabel               kidneyCancer</span></span>
<span id="cb5-15"><a href="#cb5-15" tabindex="-1"></a><span class="co">#&gt;  ---                                                            </span></span>
<span id="cb5-16"><a href="#cb5-16" tabindex="-1"></a><span class="co">#&gt; 899: tcell_McGill0107Peaks shortLabel      tcell_McGill0107Peaks</span></span>
<span id="cb5-17"><a href="#cb5-17" tabindex="-1"></a><span class="co">#&gt; 900: tcell_McGill0107Peaks  longLabel tcell | McGill0107 | Peaks</span></span>
<span id="cb5-18"><a href="#cb5-18" tabindex="-1"></a><span class="co">#&gt; 901: tcell_McGill0107Peaks     parent           tcell_McGill0107</span></span>
<span id="cb5-19"><a href="#cb5-19" tabindex="-1"></a><span class="co">#&gt; 902: tcell_McGill0107Peaks       type                     bigWig</span></span>
<span id="cb5-20"><a href="#cb5-20" tabindex="-1"></a><span class="co">#&gt; 903: tcell_McGill0107Peaks      color                      0,0,0</span></span>
<span id="cb5-21"><a href="#cb5-21" tabindex="-1"></a><span class="fu">str</span>(fields.dt)</span>
<span id="cb5-22"><a href="#cb5-22" tabindex="-1"></a><span class="co">#&gt; Classes &#39;data.table&#39; and &#39;data.frame&#39;:   903 obs. of  3 variables:</span></span>
<span id="cb5-23"><a href="#cb5-23" tabindex="-1"></a><span class="co">#&gt;  $ track   : chr  &quot;bcell&quot; &quot;bcell&quot; &quot;bcell&quot; &quot;kidneyCancer&quot; ...</span></span>
<span id="cb5-24"><a href="#cb5-24" tabindex="-1"></a><span class="co">#&gt;  $ variable: chr  &quot;superTrack&quot; &quot;shortLabel&quot; &quot;longLabel&quot; &quot;superTrack&quot; ...</span></span>
<span id="cb5-25"><a href="#cb5-25" tabindex="-1"></a><span class="co">#&gt;  $ value   : chr  &quot;on show&quot; &quot;bcell&quot; &quot;bcell ChIP-seq samples&quot; &quot;on show&quot; ...</span></span>
<span id="cb5-26"><a href="#cb5-26" tabindex="-1"></a><span class="co">#&gt;  - attr(*, &quot;.internal.selfref&quot;)=&lt;externalptr&gt;</span></span></code></pre></div>
<p>Note that because <code>by=track</code> was specified,
<code>nc::capture_all_str</code> is called for each unique value of
<code>track</code> (i.e. each row). The results are combined into a
single data.table with one row for each field. This data.table can be
easily queried, e.g.</p>
<div class="sourceCode" id="cb6"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb6-1"><a href="#cb6-1" tabindex="-1"></a>fields.dt[</span>
<span id="cb6-2"><a href="#cb6-2" tabindex="-1"></a>  <span class="fu">J</span>(<span class="st">&quot;tcell_McGill0107Coverage&quot;</span>, <span class="st">&quot;bigDataUrl&quot;</span>),</span>
<span id="cb6-3"><a href="#cb6-3" tabindex="-1"></a>  value,</span>
<span id="cb6-4"><a href="#cb6-4" tabindex="-1"></a>  on<span class="ot">=</span>.(track, variable)]</span>
<span id="cb6-5"><a href="#cb6-5" tabindex="-1"></a><span class="co">#&gt; [1] &quot;http://hubs.hpc.mcgill.ca/~thocking/PeakSegFPOP-/samples/tcell/McGill0107/coverage.bigWig&quot;</span></span>
<span id="cb6-6"><a href="#cb6-6" tabindex="-1"></a>fields.dt[, .(<span class="at">count=</span>.N), by<span class="ot">=</span>variable][<span class="fu">order</span>(count)]</span>
<span id="cb6-7"><a href="#cb6-7" tabindex="-1"></a><span class="co">#&gt;                  variable count</span></span>
<span id="cb6-8"><a href="#cb6-8" tabindex="-1"></a><span class="co">#&gt;                    &lt;char&gt; &lt;int&gt;</span></span>
<span id="cb6-9"><a href="#cb6-9" tabindex="-1"></a><span class="co">#&gt;  1:               itemRgb     4</span></span>
<span id="cb6-10"><a href="#cb6-10" tabindex="-1"></a><span class="co">#&gt;  2:              spectrum     4</span></span>
<span id="cb6-11"><a href="#cb6-11" tabindex="-1"></a><span class="co">#&gt;  3:            superTrack     8</span></span>
<span id="cb6-12"><a href="#cb6-12" tabindex="-1"></a><span class="co">#&gt;  4:             container    37</span></span>
<span id="cb6-13"><a href="#cb6-13" tabindex="-1"></a><span class="co">#&gt;  5:             graphType    37</span></span>
<span id="cb6-14"><a href="#cb6-14" tabindex="-1"></a><span class="co">#&gt;  6:             aggregate    37</span></span>
<span id="cb6-15"><a href="#cb6-15" tabindex="-1"></a><span class="co">#&gt;  7: showSubtrackColorOnUi    37</span></span>
<span id="cb6-16"><a href="#cb6-16" tabindex="-1"></a><span class="co">#&gt;  8:       maxHeightPixels    37</span></span>
<span id="cb6-17"><a href="#cb6-17" tabindex="-1"></a><span class="co">#&gt;  9:             autoScale    37</span></span>
<span id="cb6-18"><a href="#cb6-18" tabindex="-1"></a><span class="co">#&gt; 10:            visibility    41</span></span>
<span id="cb6-19"><a href="#cb6-19" tabindex="-1"></a><span class="co">#&gt; 11:                 color    74</span></span>
<span id="cb6-20"><a href="#cb6-20" tabindex="-1"></a><span class="co">#&gt; 12:            bigDataUrl    78</span></span>
<span id="cb6-21"><a href="#cb6-21" tabindex="-1"></a><span class="co">#&gt; 13:                parent   111</span></span>
<span id="cb6-22"><a href="#cb6-22" tabindex="-1"></a><span class="co">#&gt; 14:                  type   115</span></span>
<span id="cb6-23"><a href="#cb6-23" tabindex="-1"></a><span class="co">#&gt; 15:            shortLabel   123</span></span>
<span id="cb6-24"><a href="#cb6-24" tabindex="-1"></a><span class="co">#&gt; 16:             longLabel   123</span></span></code></pre></div>
<p>For more information about data.table syntax, read
<code>vignette(&quot;datatable-intro&quot;, package=&quot;data.table&quot;)</code>.</p>
</div>
<div id="match-all-tracks-and-some-fields-with-one-regex" class="section level2">
<h2>Match all tracks and some fields with one regex</h2>
<p>In the examples above we extracted all fields from all tracks (using
two regexes, one for the track, one for the field). In the example below
we extract only the track name, split into separate columns (using a
single regex for the track).</p>
<div class="sourceCode" id="cb7"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb7-1"><a href="#cb7-1" tabindex="-1"></a>cell.sample.type <span class="ot">&lt;-</span> <span class="fu">list</span>(</span>
<span id="cb7-2"><a href="#cb7-2" tabindex="-1"></a>  <span class="at">cellType=</span><span class="st">&quot;[^ ]*?&quot;</span>,</span>
<span id="cb7-3"><a href="#cb7-3" tabindex="-1"></a>  <span class="st">&quot;_&quot;</span>,</span>
<span id="cb7-4"><a href="#cb7-4" tabindex="-1"></a>  <span class="at">sampleName=</span><span class="fu">list</span>(</span>
<span id="cb7-5"><a href="#cb7-5" tabindex="-1"></a>    <span class="st">&quot;McGill&quot;</span>,</span>
<span id="cb7-6"><a href="#cb7-6" tabindex="-1"></a>    <span class="at">sampleID=</span><span class="st">&quot;[0-9]+&quot;</span>, as.integer),</span>
<span id="cb7-7"><a href="#cb7-7" tabindex="-1"></a>  <span class="at">dataType=</span><span class="st">&quot;Coverage|Peaks&quot;</span>)</span>
<span id="cb7-8"><a href="#cb7-8" tabindex="-1"></a>nc<span class="sc">::</span><span class="fu">capture_all_str</span>(trackDb.vec, cell.sample.type)</span>
<span id="cb7-9"><a href="#cb7-9" tabindex="-1"></a><span class="co">#&gt;      cellType sampleName sampleID dataType</span></span>
<span id="cb7-10"><a href="#cb7-10" tabindex="-1"></a><span class="co">#&gt;        &lt;char&gt;     &lt;char&gt;    &lt;int&gt;   &lt;char&gt;</span></span>
<span id="cb7-11"><a href="#cb7-11" tabindex="-1"></a><span class="co">#&gt;   1:    bcell McGill0091       91 Coverage</span></span>
<span id="cb7-12"><a href="#cb7-12" tabindex="-1"></a><span class="co">#&gt;   2:    bcell McGill0091       91 Coverage</span></span>
<span id="cb7-13"><a href="#cb7-13" tabindex="-1"></a><span class="co">#&gt;   3:    bcell McGill0091       91    Peaks</span></span>
<span id="cb7-14"><a href="#cb7-14" tabindex="-1"></a><span class="co">#&gt;   4:    bcell McGill0091       91    Peaks</span></span>
<span id="cb7-15"><a href="#cb7-15" tabindex="-1"></a><span class="co">#&gt;   5:    bcell McGill0322      322 Coverage</span></span>
<span id="cb7-16"><a href="#cb7-16" tabindex="-1"></a><span class="co">#&gt;  ---                                      </span></span>
<span id="cb7-17"><a href="#cb7-17" tabindex="-1"></a><span class="co">#&gt; 144:    tcell McGill0106      106    Peaks</span></span>
<span id="cb7-18"><a href="#cb7-18" tabindex="-1"></a><span class="co">#&gt; 145:    tcell McGill0107      107 Coverage</span></span>
<span id="cb7-19"><a href="#cb7-19" tabindex="-1"></a><span class="co">#&gt; 146:    tcell McGill0107      107 Coverage</span></span>
<span id="cb7-20"><a href="#cb7-20" tabindex="-1"></a><span class="co">#&gt; 147:    tcell McGill0107      107    Peaks</span></span>
<span id="cb7-21"><a href="#cb7-21" tabindex="-1"></a><span class="co">#&gt; 148:    tcell McGill0107      107    Peaks</span></span></code></pre></div>
<p>Note that the pattern above defines nested capture groups via named
lists (e.g. sampleID is a subset of sampleName). The pattern below
matches either the previously specified track pattern, or any other type
of track name:</p>
<div class="sourceCode" id="cb8"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb8-1"><a href="#cb8-1" tabindex="-1"></a>sample.or.anything <span class="ot">&lt;-</span> <span class="fu">list</span>(</span>
<span id="cb8-2"><a href="#cb8-2" tabindex="-1"></a>  cell.sample.type,</span>
<span id="cb8-3"><a href="#cb8-3" tabindex="-1"></a>  <span class="st">&quot;|&quot;</span>,</span>
<span id="cb8-4"><a href="#cb8-4" tabindex="-1"></a>  <span class="st">&quot;[^</span><span class="sc">\n</span><span class="st">]+&quot;</span>)</span>
<span id="cb8-5"><a href="#cb8-5" tabindex="-1"></a>track.pattern.old <span class="ot">&lt;-</span> <span class="fu">list</span>(</span>
<span id="cb8-6"><a href="#cb8-6" tabindex="-1"></a>  <span class="st">&quot;track &quot;</span>,</span>
<span id="cb8-7"><a href="#cb8-7" tabindex="-1"></a>  <span class="at">track=</span>sample.or.anything)</span>
<span id="cb8-8"><a href="#cb8-8" tabindex="-1"></a>nc<span class="sc">::</span><span class="fu">capture_all_str</span>(trackDb.vec, track.pattern.old)</span>
<span id="cb8-9"><a href="#cb8-9" tabindex="-1"></a><span class="co">#&gt;                         track cellType sampleName sampleID dataType</span></span>
<span id="cb8-10"><a href="#cb8-10" tabindex="-1"></a><span class="co">#&gt;                        &lt;char&gt;   &lt;char&gt;     &lt;char&gt;    &lt;int&gt;   &lt;char&gt;</span></span>
<span id="cb8-11"><a href="#cb8-11" tabindex="-1"></a><span class="co">#&gt;   1:                    bcell                           NA         </span></span>
<span id="cb8-12"><a href="#cb8-12" tabindex="-1"></a><span class="co">#&gt;   2:             kidneyCancer                           NA         </span></span>
<span id="cb8-13"><a href="#cb8-13" tabindex="-1"></a><span class="co">#&gt;   3:                   kidney                           NA         </span></span>
<span id="cb8-14"><a href="#cb8-14" tabindex="-1"></a><span class="co">#&gt;   4:   leukemiaCD19CD10BCells                           NA         </span></span>
<span id="cb8-15"><a href="#cb8-15" tabindex="-1"></a><span class="co">#&gt;   5:                 monocyte                           NA         </span></span>
<span id="cb8-16"><a href="#cb8-16" tabindex="-1"></a><span class="co">#&gt;  ---                                                               </span></span>
<span id="cb8-17"><a href="#cb8-17" tabindex="-1"></a><span class="co">#&gt; 119: tcell_McGill0106Coverage    tcell McGill0106      106 Coverage</span></span>
<span id="cb8-18"><a href="#cb8-18" tabindex="-1"></a><span class="co">#&gt; 120:    tcell_McGill0106Peaks    tcell McGill0106      106    Peaks</span></span>
<span id="cb8-19"><a href="#cb8-19" tabindex="-1"></a><span class="co">#&gt; 121:         tcell_McGill0107                           NA         </span></span>
<span id="cb8-20"><a href="#cb8-20" tabindex="-1"></a><span class="co">#&gt; 122: tcell_McGill0107Coverage    tcell McGill0107      107 Coverage</span></span>
<span id="cb8-21"><a href="#cb8-21" tabindex="-1"></a><span class="co">#&gt; 123:    tcell_McGill0107Peaks    tcell McGill0107      107    Peaks</span></span></code></pre></div>
<p>Notice the repetition of <code>track</code> in the pattern above.
This can be avoided by using the <code>nc::field</code> helper function,
which takes three arguments, that are pasted together to form a
pattern:</p>
<ul>
<li><code>field.name</code> is used as a pattern, and as the capture
group (column) name for the pattern specified in the third
argument.</li>
<li><code>between.pattern</code> is a pattern that matches between the
other two patterns.</li>
<li><code>field.pattern</code> is the pattern that matches the text to
be extracted in a capture group.</li>
</ul>
<p>The example above can thus be re-written as below, avoiding the
repetition of <code>track</code> which was present above:</p>
<div class="sourceCode" id="cb9"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb9-1"><a href="#cb9-1" tabindex="-1"></a>track.pattern <span class="ot">&lt;-</span> nc<span class="sc">::</span><span class="fu">field</span>(<span class="st">&quot;track&quot;</span>, <span class="st">&quot; &quot;</span>, sample.or.anything)</span>
<span id="cb9-2"><a href="#cb9-2" tabindex="-1"></a>nc<span class="sc">::</span><span class="fu">capture_all_str</span>(trackDb.vec, track.pattern)</span>
<span id="cb9-3"><a href="#cb9-3" tabindex="-1"></a><span class="co">#&gt;                         track cellType sampleName sampleID dataType</span></span>
<span id="cb9-4"><a href="#cb9-4" tabindex="-1"></a><span class="co">#&gt;                        &lt;char&gt;   &lt;char&gt;     &lt;char&gt;    &lt;int&gt;   &lt;char&gt;</span></span>
<span id="cb9-5"><a href="#cb9-5" tabindex="-1"></a><span class="co">#&gt;   1:                    bcell                           NA         </span></span>
<span id="cb9-6"><a href="#cb9-6" tabindex="-1"></a><span class="co">#&gt;   2:             kidneyCancer                           NA         </span></span>
<span id="cb9-7"><a href="#cb9-7" tabindex="-1"></a><span class="co">#&gt;   3:                   kidney                           NA         </span></span>
<span id="cb9-8"><a href="#cb9-8" tabindex="-1"></a><span class="co">#&gt;   4:   leukemiaCD19CD10BCells                           NA         </span></span>
<span id="cb9-9"><a href="#cb9-9" tabindex="-1"></a><span class="co">#&gt;   5:                 monocyte                           NA         </span></span>
<span id="cb9-10"><a href="#cb9-10" tabindex="-1"></a><span class="co">#&gt;  ---                                                               </span></span>
<span id="cb9-11"><a href="#cb9-11" tabindex="-1"></a><span class="co">#&gt; 119: tcell_McGill0106Coverage    tcell McGill0106      106 Coverage</span></span>
<span id="cb9-12"><a href="#cb9-12" tabindex="-1"></a><span class="co">#&gt; 120:    tcell_McGill0106Peaks    tcell McGill0106      106    Peaks</span></span>
<span id="cb9-13"><a href="#cb9-13" tabindex="-1"></a><span class="co">#&gt; 121:         tcell_McGill0107                           NA         </span></span>
<span id="cb9-14"><a href="#cb9-14" tabindex="-1"></a><span class="co">#&gt; 122: tcell_McGill0107Coverage    tcell McGill0107      107 Coverage</span></span>
<span id="cb9-15"><a href="#cb9-15" tabindex="-1"></a><span class="co">#&gt; 123:    tcell_McGill0107Peaks    tcell McGill0107      107    Peaks</span></span></code></pre></div>
<p>Finally we use <code>field</code> again to match the type column:</p>
<div class="sourceCode" id="cb10"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb10-1"><a href="#cb10-1" tabindex="-1"></a>any.lines.pattern <span class="ot">&lt;-</span> <span class="st">&quot;(?:</span><span class="sc">\n</span><span class="st">[^</span><span class="sc">\n</span><span class="st">]+)*&quot;</span></span>
<span id="cb10-2"><a href="#cb10-2" tabindex="-1"></a>nc<span class="sc">::</span><span class="fu">capture_all_str</span>(</span>
<span id="cb10-3"><a href="#cb10-3" tabindex="-1"></a>  trackDb.vec,</span>
<span id="cb10-4"><a href="#cb10-4" tabindex="-1"></a>  track.pattern,</span>
<span id="cb10-5"><a href="#cb10-5" tabindex="-1"></a>  any.lines.pattern,</span>
<span id="cb10-6"><a href="#cb10-6" tabindex="-1"></a>  <span class="st">&quot;</span><span class="sc">\\</span><span class="st">s+&quot;</span>,</span>
<span id="cb10-7"><a href="#cb10-7" tabindex="-1"></a>  nc<span class="sc">::</span><span class="fu">field</span>(<span class="st">&quot;type&quot;</span>, <span class="st">&quot; &quot;</span>, <span class="st">&quot;[^</span><span class="sc">\n</span><span class="st">]+&quot;</span>))</span>
<span id="cb10-8"><a href="#cb10-8" tabindex="-1"></a><span class="co">#&gt;                         track cellType sampleName sampleID dataType     type</span></span>
<span id="cb10-9"><a href="#cb10-9" tabindex="-1"></a><span class="co">#&gt;                        &lt;char&gt;   &lt;char&gt;     &lt;char&gt;    &lt;int&gt;   &lt;char&gt;   &lt;char&gt;</span></span>
<span id="cb10-10"><a href="#cb10-10" tabindex="-1"></a><span class="co">#&gt;   1:               all_labels                           NA          bigBed 9</span></span>
<span id="cb10-11"><a href="#cb10-11" tabindex="-1"></a><span class="co">#&gt;   2:                 problems                           NA          bigBed 3</span></span>
<span id="cb10-12"><a href="#cb10-12" tabindex="-1"></a><span class="co">#&gt;   3:            jointProblems                           NA          bigBed 3</span></span>
<span id="cb10-13"><a href="#cb10-13" tabindex="-1"></a><span class="co">#&gt;   4:            peaks_summary                           NA          bigBed 5</span></span>
<span id="cb10-14"><a href="#cb10-14" tabindex="-1"></a><span class="co">#&gt;   5:         bcell_McGill0091                           NA            bigWig</span></span>
<span id="cb10-15"><a href="#cb10-15" tabindex="-1"></a><span class="co">#&gt;  ---                                                                        </span></span>
<span id="cb10-16"><a href="#cb10-16" tabindex="-1"></a><span class="co">#&gt; 111: tcell_McGill0106Coverage    tcell McGill0106      106 Coverage   bigWig</span></span>
<span id="cb10-17"><a href="#cb10-17" tabindex="-1"></a><span class="co">#&gt; 112:    tcell_McGill0106Peaks    tcell McGill0106      106    Peaks   bigWig</span></span>
<span id="cb10-18"><a href="#cb10-18" tabindex="-1"></a><span class="co">#&gt; 113:         tcell_McGill0107                           NA            bigWig</span></span>
<span id="cb10-19"><a href="#cb10-19" tabindex="-1"></a><span class="co">#&gt; 114: tcell_McGill0107Coverage    tcell McGill0107      107 Coverage   bigWig</span></span>
<span id="cb10-20"><a href="#cb10-20" tabindex="-1"></a><span class="co">#&gt; 115:    tcell_McGill0107Peaks    tcell McGill0107      107    Peaks   bigWig</span></span></code></pre></div>
<p>Exercise for the reader (easy): modify the above regex in order to
capture the bigDataUrl field, and three additional columns (red, green,
blue) from the color field. Assume that <code>bigDataUrl</code> occurs
before <code>color</code> in each track. Note that this is a limitation
of the single regex approach — using two regex, as described in previous
sections, could extract any/all fields, even if they appear in different
orders in different tracks.</p>
<p>Exercise for the reader (hard): note that the last code block only
matches tracks which define the type field. How would you optionally
match the type field? Hint: the current <code>any.lines.pattern</code>
can match the type field.</p>
</div>
<div id="parsing-sweed-output-files" class="section level1">
<h1>Parsing SweeD output files</h1>
<p>Thanks to Marc Tollis for providing the example data used in this
section (from the SweeD bioinformatics program). Some representative
lines from one output file are shown below.</p>
<div class="sourceCode" id="cb11"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb11-1"><a href="#cb11-1" tabindex="-1"></a>info.txt.gz <span class="ot">&lt;-</span> <span class="fu">system.file</span>(</span>
<span id="cb11-2"><a href="#cb11-2" tabindex="-1"></a>  <span class="st">&quot;extdata&quot;</span>, <span class="st">&quot;SweeD_Info.txt.gz&quot;</span>, <span class="at">package=</span><span class="st">&quot;nc&quot;</span>)</span>
<span id="cb11-3"><a href="#cb11-3" tabindex="-1"></a>info.vec <span class="ot">&lt;-</span> <span class="fu">readLines</span>(info.txt.gz)</span>
<span id="cb11-4"><a href="#cb11-4" tabindex="-1"></a>info.vec[<span class="dv">20</span><span class="sc">:</span><span class="dv">50</span>]</span>
<span id="cb11-5"><a href="#cb11-5" tabindex="-1"></a><span class="co">#&gt;  [1] &quot; Total number of samples in the VCF:\t13&quot;</span></span>
<span id="cb11-6"><a href="#cb11-6" tabindex="-1"></a><span class="co">#&gt;  [2] &quot; Samples excluded from the analysis:\t6&quot; </span></span>
<span id="cb11-7"><a href="#cb11-7" tabindex="-1"></a><span class="co">#&gt;  [3] &quot;&quot;                                        </span></span>
<span id="cb11-8"><a href="#cb11-8" tabindex="-1"></a><span class="co">#&gt;  [4] &quot;&quot;                                        </span></span>
<span id="cb11-9"><a href="#cb11-9" tabindex="-1"></a><span class="co">#&gt;  [5] &quot; Alignment 1&quot;                            </span></span>
<span id="cb11-10"><a href="#cb11-10" tabindex="-1"></a><span class="co">#&gt;  [6] &quot;&quot;                                        </span></span>
<span id="cb11-11"><a href="#cb11-11" tabindex="-1"></a><span class="co">#&gt;  [7] &quot;\t\tChromosome:\t\tscaffold_0&quot;           </span></span>
<span id="cb11-12"><a href="#cb11-12" tabindex="-1"></a><span class="co">#&gt;  [8] &quot;\t\tSequences:\t\t14&quot;                    </span></span>
<span id="cb11-13"><a href="#cb11-13" tabindex="-1"></a><span class="co">#&gt;  [9] &quot;\t\tSites:\t\t\t1670366&quot;                 </span></span>
<span id="cb11-14"><a href="#cb11-14" tabindex="-1"></a><span class="co">#&gt; [10] &quot;\t\tDiscarded sites:\t1264068&quot;           </span></span>
<span id="cb11-15"><a href="#cb11-15" tabindex="-1"></a><span class="co">#&gt; [11] &quot;&quot;                                        </span></span>
<span id="cb11-16"><a href="#cb11-16" tabindex="-1"></a><span class="co">#&gt; [12] &quot;\t\tProcessing:\t\t155.53 seconds&quot;       </span></span>
<span id="cb11-17"><a href="#cb11-17" tabindex="-1"></a><span class="co">#&gt; [13] &quot;&quot;                                        </span></span>
<span id="cb11-18"><a href="#cb11-18" tabindex="-1"></a><span class="co">#&gt; [14] &quot;\t\tPosition:\t\t8.936200e+07&quot;           </span></span>
<span id="cb11-19"><a href="#cb11-19" tabindex="-1"></a><span class="co">#&gt; [15] &quot;\t\tLikelihood:\t\t4.105582e+02&quot;         </span></span>
<span id="cb11-20"><a href="#cb11-20" tabindex="-1"></a><span class="co">#&gt; [16] &quot;\t\tAlpha:\t\t\t6.616326e-06&quot;            </span></span>
<span id="cb11-21"><a href="#cb11-21" tabindex="-1"></a><span class="co">#&gt; [17] &quot;&quot;                                        </span></span>
<span id="cb11-22"><a href="#cb11-22" tabindex="-1"></a><span class="co">#&gt; [18] &quot;&quot;                                        </span></span>
<span id="cb11-23"><a href="#cb11-23" tabindex="-1"></a><span class="co">#&gt; [19] &quot; Alignment 2&quot;                            </span></span>
<span id="cb11-24"><a href="#cb11-24" tabindex="-1"></a><span class="co">#&gt; [20] &quot;&quot;                                        </span></span>
<span id="cb11-25"><a href="#cb11-25" tabindex="-1"></a><span class="co">#&gt; [21] &quot;\t\tChromosome:\t\tscaffold_1&quot;           </span></span>
<span id="cb11-26"><a href="#cb11-26" tabindex="-1"></a><span class="co">#&gt; [22] &quot;\t\tSequences:\t\t14&quot;                    </span></span>
<span id="cb11-27"><a href="#cb11-27" tabindex="-1"></a><span class="co">#&gt; [23] &quot;\t\tSites:\t\t\t1447008&quot;                 </span></span>
<span id="cb11-28"><a href="#cb11-28" tabindex="-1"></a><span class="co">#&gt; [24] &quot;\t\tDiscarded sites:\t1093595&quot;           </span></span>
<span id="cb11-29"><a href="#cb11-29" tabindex="-1"></a><span class="co">#&gt; [25] &quot;&quot;                                        </span></span>
<span id="cb11-30"><a href="#cb11-30" tabindex="-1"></a><span class="co">#&gt; [26] &quot;\t\tProcessing:\t\t138.83 seconds&quot;       </span></span>
<span id="cb11-31"><a href="#cb11-31" tabindex="-1"></a><span class="co">#&gt; [27] &quot;&quot;                                        </span></span>
<span id="cb11-32"><a href="#cb11-32" tabindex="-1"></a><span class="co">#&gt; [28] &quot;\t\tPosition:\t\t8.722482e+07&quot;           </span></span>
<span id="cb11-33"><a href="#cb11-33" tabindex="-1"></a><span class="co">#&gt; [29] &quot;\t\tLikelihood:\t\t2.531514e+02&quot;         </span></span>
<span id="cb11-34"><a href="#cb11-34" tabindex="-1"></a><span class="co">#&gt; [30] &quot;\t\tAlpha:\t\t\t1.031963e-05&quot;            </span></span>
<span id="cb11-35"><a href="#cb11-35" tabindex="-1"></a><span class="co">#&gt; [31] &quot;&quot;</span></span></code></pre></div>
<p>The Alignment numbers must be matched with the numbers before slashes
in the other file,</p>
<div class="sourceCode" id="cb12"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb12-1"><a href="#cb12-1" tabindex="-1"></a>report.txt.gz <span class="ot">&lt;-</span> <span class="fu">system.file</span>(</span>
<span id="cb12-2"><a href="#cb12-2" tabindex="-1"></a>  <span class="st">&quot;extdata&quot;</span>, <span class="st">&quot;SweeD_Report.txt.gz&quot;</span>, <span class="at">package=</span><span class="st">&quot;nc&quot;</span>)</span>
<span id="cb12-3"><a href="#cb12-3" tabindex="-1"></a>report.vec <span class="ot">&lt;-</span> <span class="fu">readLines</span>(report.txt.gz)</span>
<span id="cb12-4"><a href="#cb12-4" tabindex="-1"></a><span class="fu">cat</span>(report.vec[<span class="dv">1</span><span class="sc">:</span><span class="dv">10</span>], <span class="at">sep=</span><span class="st">&quot;</span><span class="sc">\n</span><span class="st">&quot;</span>)</span>
<span id="cb12-5"><a href="#cb12-5" tabindex="-1"></a><span class="co">#&gt; </span></span>
<span id="cb12-6"><a href="#cb12-6" tabindex="-1"></a><span class="co">#&gt; //1</span></span>
<span id="cb12-7"><a href="#cb12-7" tabindex="-1"></a><span class="co">#&gt; Position Likelihood  Alpha</span></span>
<span id="cb12-8"><a href="#cb12-8" tabindex="-1"></a><span class="co">#&gt; 700.0000 4.637328e-03    2.763840e+02</span></span>
<span id="cb12-9"><a href="#cb12-9" tabindex="-1"></a><span class="co">#&gt; 130585.6172  3.781283e-01    8.490200e-04</span></span>
<span id="cb12-10"><a href="#cb12-10" tabindex="-1"></a><span class="co">#&gt; 260471.2344  3.602315e-02    4.691340e-03</span></span>
<span id="cb12-11"><a href="#cb12-11" tabindex="-1"></a><span class="co">#&gt; 390356.8516  7.618749e-01    5.377668e-04</span></span>
<span id="cb12-12"><a href="#cb12-12" tabindex="-1"></a><span class="co">#&gt; 520242.4688  2.979971e-08    1.411765e-01</span></span>
<span id="cb12-13"><a href="#cb12-13" tabindex="-1"></a><span class="co">#&gt; 650128.0859  3.552965e-03    7.790821e-03</span></span>
<span id="cb12-14"><a href="#cb12-14" tabindex="-1"></a><span class="co">#&gt; 780013.7031  4.637359e-03    1.727400e-02</span></span>
<span id="cb12-15"><a href="#cb12-15" tabindex="-1"></a><span class="fu">cat</span>(report.vec[<span class="dv">1000</span><span class="sc">:</span><span class="dv">1010</span>], <span class="at">sep=</span><span class="st">&quot;</span><span class="sc">\n</span><span class="st">&quot;</span>)</span>
<span id="cb12-16"><a href="#cb12-16" tabindex="-1"></a><span class="co">#&gt; 129366774.7188   1.218965e-01    2.215489e-02</span></span>
<span id="cb12-17"><a href="#cb12-17" tabindex="-1"></a><span class="co">#&gt; 129496660.3359   1.165627e-02    3.384931e-02</span></span>
<span id="cb12-18"><a href="#cb12-18" tabindex="-1"></a><span class="co">#&gt; 129626545.9531   2.233934e-02    3.602669e-02</span></span>
<span id="cb12-19"><a href="#cb12-19" tabindex="-1"></a><span class="co">#&gt; 129756434.0000   3.850623e-01    4.812648e+01</span></span>
<span id="cb12-20"><a href="#cb12-20" tabindex="-1"></a><span class="co">#&gt; </span></span>
<span id="cb12-21"><a href="#cb12-21" tabindex="-1"></a><span class="co">#&gt; //2</span></span>
<span id="cb12-22"><a href="#cb12-22" tabindex="-1"></a><span class="co">#&gt; Position Likelihood  Alpha</span></span>
<span id="cb12-23"><a href="#cb12-23" tabindex="-1"></a><span class="co">#&gt; 135.0000 7.282316e-01    3.163686e+01</span></span>
<span id="cb12-24"><a href="#cb12-24" tabindex="-1"></a><span class="co">#&gt; 111533.0625  2.548831e+00    4.932014e-04</span></span>
<span id="cb12-25"><a href="#cb12-25" tabindex="-1"></a><span class="co">#&gt; 222931.1250  1.369720e-02    1.044774e+00</span></span>
<span id="cb12-26"><a href="#cb12-26" tabindex="-1"></a><span class="co">#&gt; 334329.1875  7.118828e+00    3.965791e-04</span></span></code></pre></div>
<p>The goal is to produce a bed file, which has tab-separated values
with four columns: chrom, chromStart, chromEnd, Likelihood. The chrom
values appear in the info file (Chromosome) so we will need to join the
two files based on alignment ID. First we capture all alignments in the
info file:</p>
<div class="sourceCode" id="cb13"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb13-1"><a href="#cb13-1" tabindex="-1"></a>(info.dt <span class="ot">&lt;-</span> nc<span class="sc">::</span><span class="fu">capture_all_str</span>(</span>
<span id="cb13-2"><a href="#cb13-2" tabindex="-1"></a>  info.vec,</span>
<span id="cb13-3"><a href="#cb13-3" tabindex="-1"></a>  <span class="st">&quot;Alignment &quot;</span>,</span>
<span id="cb13-4"><a href="#cb13-4" tabindex="-1"></a>  <span class="at">alignment=</span><span class="st">&quot;[0-9]+&quot;</span>,</span>
<span id="cb13-5"><a href="#cb13-5" tabindex="-1"></a>  <span class="st">&quot;</span><span class="sc">\n\n\t\t</span><span class="st">Chromosome:</span><span class="sc">\t\t</span><span class="st">&quot;</span>,</span>
<span id="cb13-6"><a href="#cb13-6" tabindex="-1"></a>  <span class="at">chrom=</span><span class="st">&quot;.*&quot;</span>,</span>
<span id="cb13-7"><a href="#cb13-7" tabindex="-1"></a>  <span class="st">&quot;</span><span class="sc">\n</span><span class="st">&quot;</span>))</span>
<span id="cb13-8"><a href="#cb13-8" tabindex="-1"></a><span class="co">#&gt;     alignment      chrom</span></span>
<span id="cb13-9"><a href="#cb13-9" tabindex="-1"></a><span class="co">#&gt;        &lt;char&gt;     &lt;char&gt;</span></span>
<span id="cb13-10"><a href="#cb13-10" tabindex="-1"></a><span class="co">#&gt;  1:         1 scaffold_0</span></span>
<span id="cb13-11"><a href="#cb13-11" tabindex="-1"></a><span class="co">#&gt;  2:         2 scaffold_1</span></span>
<span id="cb13-12"><a href="#cb13-12" tabindex="-1"></a><span class="co">#&gt;  3:         3 scaffold_2</span></span>
<span id="cb13-13"><a href="#cb13-13" tabindex="-1"></a><span class="co">#&gt;  4:         4 scaffold_3</span></span>
<span id="cb13-14"><a href="#cb13-14" tabindex="-1"></a><span class="co">#&gt;  5:         5 scaffold_4</span></span>
<span id="cb13-15"><a href="#cb13-15" tabindex="-1"></a><span class="co">#&gt;  6:         6 scaffold_5</span></span>
<span id="cb13-16"><a href="#cb13-16" tabindex="-1"></a><span class="co">#&gt;  7:         7 scaffold_6</span></span>
<span id="cb13-17"><a href="#cb13-17" tabindex="-1"></a><span class="co">#&gt;  8:         8 scaffold_7</span></span>
<span id="cb13-18"><a href="#cb13-18" tabindex="-1"></a><span class="co">#&gt;  9:         9 scaffold_8</span></span>
<span id="cb13-19"><a href="#cb13-19" tabindex="-1"></a><span class="co">#&gt; 10:        10 scaffold_9</span></span></code></pre></div>
<p>Then we capture all alignment/csv blocks in the report file:</p>
<div class="sourceCode" id="cb14"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb14-1"><a href="#cb14-1" tabindex="-1"></a>(report.dt <span class="ot">&lt;-</span> nc<span class="sc">::</span><span class="fu">capture_all_str</span>(</span>
<span id="cb14-2"><a href="#cb14-2" tabindex="-1"></a>  report.vec,</span>
<span id="cb14-3"><a href="#cb14-3" tabindex="-1"></a>  <span class="st">&quot;//&quot;</span>,</span>
<span id="cb14-4"><a href="#cb14-4" tabindex="-1"></a>  <span class="at">alignment=</span><span class="st">&quot;[0-9]+&quot;</span>,</span>
<span id="cb14-5"><a href="#cb14-5" tabindex="-1"></a>  <span class="st">&quot;</span><span class="sc">\n</span><span class="st">&quot;</span>,</span>
<span id="cb14-6"><a href="#cb14-6" tabindex="-1"></a>  <span class="at">csv=</span><span class="st">&quot;[^/]+&quot;</span></span>
<span id="cb14-7"><a href="#cb14-7" tabindex="-1"></a>)[, {</span>
<span id="cb14-8"><a href="#cb14-8" tabindex="-1"></a>  data.table<span class="sc">::</span><span class="fu">fread</span>(<span class="at">text=</span>csv)</span>
<span id="cb14-9"><a href="#cb14-9" tabindex="-1"></a>}, <span class="at">by=</span>alignment])</span>
<span id="cb14-10"><a href="#cb14-10" tabindex="-1"></a><span class="co">#&gt;        alignment   Position   Likelihood        Alpha</span></span>
<span id="cb14-11"><a href="#cb14-11" tabindex="-1"></a><span class="co">#&gt;           &lt;char&gt;      &lt;num&gt;        &lt;num&gt;        &lt;num&gt;</span></span>
<span id="cb14-12"><a href="#cb14-12" tabindex="-1"></a><span class="co">#&gt;     1:         1      700.0 4.637328e-03 2.763840e+02</span></span>
<span id="cb14-13"><a href="#cb14-13" tabindex="-1"></a><span class="co">#&gt;     2:         1   130585.6 3.781283e-01 8.490200e-04</span></span>
<span id="cb14-14"><a href="#cb14-14" tabindex="-1"></a><span class="co">#&gt;     3:         1   260471.2 3.602315e-02 4.691340e-03</span></span>
<span id="cb14-15"><a href="#cb14-15" tabindex="-1"></a><span class="co">#&gt;     4:         1   390356.9 7.618749e-01 5.377668e-04</span></span>
<span id="cb14-16"><a href="#cb14-16" tabindex="-1"></a><span class="co">#&gt;     5:         1   520242.5 2.979971e-08 1.411765e-01</span></span>
<span id="cb14-17"><a href="#cb14-17" tabindex="-1"></a><span class="co">#&gt;    ---                                               </span></span>
<span id="cb14-18"><a href="#cb14-18" tabindex="-1"></a><span class="co">#&gt;  9996:        10 82991564.8 8.051006e-03 1.357819e-03</span></span>
<span id="cb14-19"><a href="#cb14-19" tabindex="-1"></a><span class="co">#&gt;  9997:        10 83074967.8 7.048433e-03 1.825764e-03</span></span>
<span id="cb14-20"><a href="#cb14-20" tabindex="-1"></a><span class="co">#&gt;  9998:        10 83158370.8 1.012360e-07 7.999999e-03</span></span>
<span id="cb14-21"><a href="#cb14-21" tabindex="-1"></a><span class="co">#&gt;  9999:        10 83241773.8 3.977189e-08 9.999997e-01</span></span>
<span id="cb14-22"><a href="#cb14-22" tabindex="-1"></a><span class="co">#&gt; 10000:        10 83325174.0 3.980538e-08 1.200000e+03</span></span></code></pre></div>
<p>Note that because <code>by=alignment</code> was specified,
<code>fread</code> is called for each unique value of
<code>alignment</code> (i.e. each row). The results are combined into a
single data.table with all of the csv data from the original file, plus
the additional <code>alignment</code> column. Next, we join this table
to the previous table in order to get the <code>chrom</code> column:</p>
<div class="sourceCode" id="cb15"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb15-1"><a href="#cb15-1" tabindex="-1"></a>(join.dt <span class="ot">&lt;-</span> report.dt[info.dt, <span class="at">on=</span>.(alignment)])</span>
<span id="cb15-2"><a href="#cb15-2" tabindex="-1"></a><span class="co">#&gt;        alignment   Position   Likelihood        Alpha      chrom</span></span>
<span id="cb15-3"><a href="#cb15-3" tabindex="-1"></a><span class="co">#&gt;           &lt;char&gt;      &lt;num&gt;        &lt;num&gt;        &lt;num&gt;     &lt;char&gt;</span></span>
<span id="cb15-4"><a href="#cb15-4" tabindex="-1"></a><span class="co">#&gt;     1:         1      700.0 4.637328e-03 2.763840e+02 scaffold_0</span></span>
<span id="cb15-5"><a href="#cb15-5" tabindex="-1"></a><span class="co">#&gt;     2:         1   130585.6 3.781283e-01 8.490200e-04 scaffold_0</span></span>
<span id="cb15-6"><a href="#cb15-6" tabindex="-1"></a><span class="co">#&gt;     3:         1   260471.2 3.602315e-02 4.691340e-03 scaffold_0</span></span>
<span id="cb15-7"><a href="#cb15-7" tabindex="-1"></a><span class="co">#&gt;     4:         1   390356.9 7.618749e-01 5.377668e-04 scaffold_0</span></span>
<span id="cb15-8"><a href="#cb15-8" tabindex="-1"></a><span class="co">#&gt;     5:         1   520242.5 2.979971e-08 1.411765e-01 scaffold_0</span></span>
<span id="cb15-9"><a href="#cb15-9" tabindex="-1"></a><span class="co">#&gt;    ---                                                          </span></span>
<span id="cb15-10"><a href="#cb15-10" tabindex="-1"></a><span class="co">#&gt;  9996:        10 82991564.8 8.051006e-03 1.357819e-03 scaffold_9</span></span>
<span id="cb15-11"><a href="#cb15-11" tabindex="-1"></a><span class="co">#&gt;  9997:        10 83074967.8 7.048433e-03 1.825764e-03 scaffold_9</span></span>
<span id="cb15-12"><a href="#cb15-12" tabindex="-1"></a><span class="co">#&gt;  9998:        10 83158370.8 1.012360e-07 7.999999e-03 scaffold_9</span></span>
<span id="cb15-13"><a href="#cb15-13" tabindex="-1"></a><span class="co">#&gt;  9999:        10 83241773.8 3.977189e-08 9.999997e-01 scaffold_9</span></span>
<span id="cb15-14"><a href="#cb15-14" tabindex="-1"></a><span class="co">#&gt; 10000:        10 83325174.0 3.980538e-08 1.200000e+03 scaffold_9</span></span></code></pre></div>
<p>Finally the desired bed table can be created via</p>
<div class="sourceCode" id="cb16"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb16-1"><a href="#cb16-1" tabindex="-1"></a>join.dt[, .(</span>
<span id="cb16-2"><a href="#cb16-2" tabindex="-1"></a>  chrom,</span>
<span id="cb16-3"><a href="#cb16-3" tabindex="-1"></a>  <span class="at">chromStart=</span><span class="fu">as.integer</span>(Position<span class="dv">-1</span>),</span>
<span id="cb16-4"><a href="#cb16-4" tabindex="-1"></a>  <span class="at">chromEnd=</span><span class="fu">as.integer</span>(Position),</span>
<span id="cb16-5"><a href="#cb16-5" tabindex="-1"></a>  Likelihood)]</span>
<span id="cb16-6"><a href="#cb16-6" tabindex="-1"></a><span class="co">#&gt;             chrom chromStart chromEnd   Likelihood</span></span>
<span id="cb16-7"><a href="#cb16-7" tabindex="-1"></a><span class="co">#&gt;            &lt;char&gt;      &lt;int&gt;    &lt;int&gt;        &lt;num&gt;</span></span>
<span id="cb16-8"><a href="#cb16-8" tabindex="-1"></a><span class="co">#&gt;     1: scaffold_0        699      700 4.637328e-03</span></span>
<span id="cb16-9"><a href="#cb16-9" tabindex="-1"></a><span class="co">#&gt;     2: scaffold_0     130584   130585 3.781283e-01</span></span>
<span id="cb16-10"><a href="#cb16-10" tabindex="-1"></a><span class="co">#&gt;     3: scaffold_0     260470   260471 3.602315e-02</span></span>
<span id="cb16-11"><a href="#cb16-11" tabindex="-1"></a><span class="co">#&gt;     4: scaffold_0     390355   390356 7.618749e-01</span></span>
<span id="cb16-12"><a href="#cb16-12" tabindex="-1"></a><span class="co">#&gt;     5: scaffold_0     520241   520242 2.979971e-08</span></span>
<span id="cb16-13"><a href="#cb16-13" tabindex="-1"></a><span class="co">#&gt;    ---                                            </span></span>
<span id="cb16-14"><a href="#cb16-14" tabindex="-1"></a><span class="co">#&gt;  9996: scaffold_9   82991563 82991564 8.051006e-03</span></span>
<span id="cb16-15"><a href="#cb16-15" tabindex="-1"></a><span class="co">#&gt;  9997: scaffold_9   83074966 83074967 7.048433e-03</span></span>
<span id="cb16-16"><a href="#cb16-16" tabindex="-1"></a><span class="co">#&gt;  9998: scaffold_9   83158369 83158370 1.012360e-07</span></span>
<span id="cb16-17"><a href="#cb16-17" tabindex="-1"></a><span class="co">#&gt;  9999: scaffold_9   83241772 83241773 3.977189e-08</span></span>
<span id="cb16-18"><a href="#cb16-18" tabindex="-1"></a><span class="co">#&gt; 10000: scaffold_9   83325173 83325174 3.980538e-08</span></span></code></pre></div>
<p>Exercise for the reader (easy): notice that the code above for
creating <code>info.dt</code> involves repetition in the pattern and
group names (<code>alignment</code>, <code>Alignment</code>,
<code>chrom</code>, <code>Chromosome</code>). Re-write the pattern using
<code>nc::field</code> in order to eliminate that repetition.</p>
<p>Exercise for the reader (hard): notice that Chromosome is only the
first field – how could you extract the other fields as well? Hint: use
<code>nc::field</code> in a helper function in order to avoid
repetition.</p>
</div>



<!-- code folding -->


<!-- dynamically load mathjax for compatibility with self-contained -->
<script>
  (function () {
    var script = document.createElement("script");
    script.type = "text/javascript";
    script.src  = "https://mathjax.rstudio.com/latest/MathJax.js?config=TeX-AMS-MML_HTMLorMML";
    document.getElementsByTagName("head")[0].appendChild(script);
  })();
</script>

</body>
</html>
