<!DOCTYPE html>

<html>

<head>

<meta charset="utf-8" />
<meta name="generator" content="pandoc" />
<meta http-equiv="X-UA-Compatible" content="IE=EDGE" />

<meta name="viewport" content="width=device-width, initial-scale=1" />


<meta name="date" content="2025-04-08" />

<title>Reading regularly named files</title>

<script>// Pandoc 2.9 adds attributes on both header and div. We remove the former (to
// be compatible with the behavior of Pandoc < 2.8).
document.addEventListener('DOMContentLoaded', function(e) {
  var hs = document.querySelectorAll("div.section[class*='level'] > :first-child");
  var i, h, a;
  for (i = 0; i < hs.length; i++) {
    h = hs[i];
    if (!/^h[1-6]$/i.test(h.tagName)) continue;  // it should be a header h1-h6
    a = h.attributes;
    while (a.length > 0) h.removeAttribute(a[0].name);
  }
});
</script>

<style type="text/css">
code{white-space: pre-wrap;}
span.smallcaps{font-variant: small-caps;}
span.underline{text-decoration: underline;}
div.column{display: inline-block; vertical-align: top; width: 50%;}
div.hanging-indent{margin-left: 1.5em; text-indent: -1.5em;}
ul.task-list{list-style: none;}
</style>



<style type="text/css">
code {
white-space: pre;
}
.sourceCode {
overflow: visible;
}
</style>
<style type="text/css" data-origin="pandoc">
pre > code.sourceCode { white-space: pre; position: relative; }
pre > code.sourceCode > span { line-height: 1.25; }
pre > code.sourceCode > span:empty { height: 1.2em; }
.sourceCode { overflow: visible; }
code.sourceCode > span { color: inherit; text-decoration: inherit; }
div.sourceCode { margin: 1em 0; }
pre.sourceCode { margin: 0; }
@media screen {
div.sourceCode { overflow: auto; }
}
@media print {
pre > code.sourceCode { white-space: pre-wrap; }
pre > code.sourceCode > span { text-indent: -5em; padding-left: 5em; }
}
pre.numberSource code
{ counter-reset: source-line 0; }
pre.numberSource code > span
{ position: relative; left: -4em; counter-increment: source-line; }
pre.numberSource code > span > a:first-child::before
{ content: counter(source-line);
position: relative; left: -1em; text-align: right; vertical-align: baseline;
border: none; display: inline-block;
-webkit-touch-callout: none; -webkit-user-select: none;
-khtml-user-select: none; -moz-user-select: none;
-ms-user-select: none; user-select: none;
padding: 0 4px; width: 4em;
color: #aaaaaa;
}
pre.numberSource { margin-left: 3em; border-left: 1px solid #aaaaaa; padding-left: 4px; }
div.sourceCode
{ }
@media screen {
pre > code.sourceCode > span > a:first-child::before { text-decoration: underline; }
}
code span.al { color: #ff0000; font-weight: bold; } 
code span.an { color: #60a0b0; font-weight: bold; font-style: italic; } 
code span.at { color: #7d9029; } 
code span.bn { color: #40a070; } 
code span.bu { color: #008000; } 
code span.cf { color: #007020; font-weight: bold; } 
code span.ch { color: #4070a0; } 
code span.cn { color: #880000; } 
code span.co { color: #60a0b0; font-style: italic; } 
code span.cv { color: #60a0b0; font-weight: bold; font-style: italic; } 
code span.do { color: #ba2121; font-style: italic; } 
code span.dt { color: #902000; } 
code span.dv { color: #40a070; } 
code span.er { color: #ff0000; font-weight: bold; } 
code span.ex { } 
code span.fl { color: #40a070; } 
code span.fu { color: #06287e; } 
code span.im { color: #008000; font-weight: bold; } 
code span.in { color: #60a0b0; font-weight: bold; font-style: italic; } 
code span.kw { color: #007020; font-weight: bold; } 
code span.op { color: #666666; } 
code span.ot { color: #007020; } 
code span.pp { color: #bc7a00; } 
code span.sc { color: #4070a0; } 
code span.ss { color: #bb6688; } 
code span.st { color: #4070a0; } 
code span.va { color: #19177c; } 
code span.vs { color: #4070a0; } 
code span.wa { color: #60a0b0; font-weight: bold; font-style: italic; } 
</style>
<script>
// apply pandoc div.sourceCode style to pre.sourceCode instead
(function() {
  var sheets = document.styleSheets;
  for (var i = 0; i < sheets.length; i++) {
    if (sheets[i].ownerNode.dataset["origin"] !== "pandoc") continue;
    try { var rules = sheets[i].cssRules; } catch (e) { continue; }
    var j = 0;
    while (j < rules.length) {
      var rule = rules[j];
      // check if there is a div.sourceCode rule
      if (rule.type !== rule.STYLE_RULE || rule.selectorText !== "div.sourceCode") {
        j++;
        continue;
      }
      var style = rule.style.cssText;
      // check if color or background-color is set
      if (rule.style.color === '' && rule.style.backgroundColor === '') {
        j++;
        continue;
      }
      // replace div.sourceCode by a pre.sourceCode rule
      sheets[i].deleteRule(j);
      sheets[i].insertRule('pre.sourceCode{' + style + '}', j);
    }
  }
})();
</script>




<style type="text/css">body {
background-color: #fff;
margin: 1em auto;
max-width: 700px;
overflow: visible;
padding-left: 2em;
padding-right: 2em;
font-family: "Open Sans", "Helvetica Neue", Helvetica, Arial, sans-serif;
font-size: 14px;
line-height: 1.35;
}
#TOC {
clear: both;
margin: 0 0 10px 10px;
padding: 4px;
width: 400px;
border: 1px solid #CCCCCC;
border-radius: 5px;
background-color: #f6f6f6;
font-size: 13px;
line-height: 1.3;
}
#TOC .toctitle {
font-weight: bold;
font-size: 15px;
margin-left: 5px;
}
#TOC ul {
padding-left: 40px;
margin-left: -1.5em;
margin-top: 5px;
margin-bottom: 5px;
}
#TOC ul ul {
margin-left: -2em;
}
#TOC li {
line-height: 16px;
}
table {
margin: 1em auto;
border-width: 1px;
border-color: #DDDDDD;
border-style: outset;
border-collapse: collapse;
}
table th {
border-width: 2px;
padding: 5px;
border-style: inset;
}
table td {
border-width: 1px;
border-style: inset;
line-height: 18px;
padding: 5px 5px;
}
table, table th, table td {
border-left-style: none;
border-right-style: none;
}
table thead, table tr.even {
background-color: #f7f7f7;
}
p {
margin: 0.5em 0;
}
blockquote {
background-color: #f6f6f6;
padding: 0.25em 0.75em;
}
hr {
border-style: solid;
border: none;
border-top: 1px solid #777;
margin: 28px 0;
}
dl {
margin-left: 0;
}
dl dd {
margin-bottom: 13px;
margin-left: 13px;
}
dl dt {
font-weight: bold;
}
ul {
margin-top: 0;
}
ul li {
list-style: circle outside;
}
ul ul {
margin-bottom: 0;
}
pre, code {
background-color: #f7f7f7;
border-radius: 3px;
color: #333;
white-space: pre-wrap; 
}
pre {
border-radius: 3px;
margin: 5px 0px 10px 0px;
padding: 10px;
}
pre:not([class]) {
background-color: #f7f7f7;
}
code {
font-family: Consolas, Monaco, 'Courier New', monospace;
font-size: 85%;
}
p > code, li > code {
padding: 2px 0px;
}
div.figure {
text-align: center;
}
img {
background-color: #FFFFFF;
padding: 2px;
border: 1px solid #DDDDDD;
border-radius: 3px;
border: 1px solid #CCCCCC;
margin: 0 5px;
}
h1 {
margin-top: 0;
font-size: 35px;
line-height: 40px;
}
h2 {
border-bottom: 4px solid #f7f7f7;
padding-top: 10px;
padding-bottom: 2px;
font-size: 145%;
}
h3 {
border-bottom: 2px solid #f7f7f7;
padding-top: 10px;
font-size: 120%;
}
h4 {
border-bottom: 1px solid #f7f7f7;
margin-left: 8px;
font-size: 105%;
}
h5, h6 {
border-bottom: 1px solid #ccc;
font-size: 105%;
}
a {
color: #0033dd;
text-decoration: none;
}
a:hover {
color: #6666ff; }
a:visited {
color: #800080; }
a:visited:hover {
color: #BB00BB; }
a[href^="http:"] {
text-decoration: underline; }
a[href^="https:"] {
text-decoration: underline; }

code > span.kw { color: #555; font-weight: bold; } 
code > span.dt { color: #902000; } 
code > span.dv { color: #40a070; } 
code > span.bn { color: #d14; } 
code > span.fl { color: #d14; } 
code > span.ch { color: #d14; } 
code > span.st { color: #d14; } 
code > span.co { color: #888888; font-style: italic; } 
code > span.ot { color: #007020; } 
code > span.al { color: #ff0000; font-weight: bold; } 
code > span.fu { color: #900; font-weight: bold; } 
code > span.er { color: #a61717; background-color: #e3d2d2; } 
</style>




</head>

<body>




<h1 class="title toc-ignore">Reading regularly named files</h1>
<h4 class="date">2025-04-08</h4>



<p>This vignette contains a number of examples which explain how to use
<code>capture_first_glob</code> to read data from a set of regularly
named files.</p>
<div id="example-0-iris-data-one-file-per-species" class="section level2">
<h2>Example 0: iris data, one file per species</h2>
<p>We begin with a simple example: iris data have 150 rows, as shown
below.</p>
<div class="sourceCode" id="cb1"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb1-1"><a href="#cb1-1" tabindex="-1"></a><span class="fu">library</span>(data.table)</span>
<span id="cb1-2"><a href="#cb1-2" tabindex="-1"></a><span class="fu">dir.create</span>(iris.dir <span class="ot">&lt;-</span> <span class="fu">tempfile</span>())</span>
<span id="cb1-3"><a href="#cb1-3" tabindex="-1"></a>icsv <span class="ot">&lt;-</span> <span class="cf">function</span>(sp)<span class="fu">file.path</span>(iris.dir, <span class="fu">paste0</span>(sp, <span class="st">&quot;.csv&quot;</span>))</span>
<span id="cb1-4"><a href="#cb1-4" tabindex="-1"></a>(iris.dt <span class="ot">&lt;-</span> <span class="fu">data.table</span>(iris))</span>
<span id="cb1-5"><a href="#cb1-5" tabindex="-1"></a><span class="co">#&gt;      Sepal.Length Sepal.Width Petal.Length Petal.Width   Species</span></span>
<span id="cb1-6"><a href="#cb1-6" tabindex="-1"></a><span class="co">#&gt;             &lt;num&gt;       &lt;num&gt;        &lt;num&gt;       &lt;num&gt;    &lt;fctr&gt;</span></span>
<span id="cb1-7"><a href="#cb1-7" tabindex="-1"></a><span class="co">#&gt;   1:          5.1         3.5          1.4         0.2    setosa</span></span>
<span id="cb1-8"><a href="#cb1-8" tabindex="-1"></a><span class="co">#&gt;   2:          4.9         3.0          1.4         0.2    setosa</span></span>
<span id="cb1-9"><a href="#cb1-9" tabindex="-1"></a><span class="co">#&gt;   3:          4.7         3.2          1.3         0.2    setosa</span></span>
<span id="cb1-10"><a href="#cb1-10" tabindex="-1"></a><span class="co">#&gt;   4:          4.6         3.1          1.5         0.2    setosa</span></span>
<span id="cb1-11"><a href="#cb1-11" tabindex="-1"></a><span class="co">#&gt;   5:          5.0         3.6          1.4         0.2    setosa</span></span>
<span id="cb1-12"><a href="#cb1-12" tabindex="-1"></a><span class="co">#&gt;  ---                                                            </span></span>
<span id="cb1-13"><a href="#cb1-13" tabindex="-1"></a><span class="co">#&gt; 146:          6.7         3.0          5.2         2.3 virginica</span></span>
<span id="cb1-14"><a href="#cb1-14" tabindex="-1"></a><span class="co">#&gt; 147:          6.3         2.5          5.0         1.9 virginica</span></span>
<span id="cb1-15"><a href="#cb1-15" tabindex="-1"></a><span class="co">#&gt; 148:          6.5         3.0          5.2         2.0 virginica</span></span>
<span id="cb1-16"><a href="#cb1-16" tabindex="-1"></a><span class="co">#&gt; 149:          6.2         3.4          5.4         2.3 virginica</span></span>
<span id="cb1-17"><a href="#cb1-17" tabindex="-1"></a><span class="co">#&gt; 150:          5.9         3.0          5.1         1.8 virginica</span></span></code></pre></div>
<p>In the code below, we save one CSV file for each of the three
Species.</p>
<div class="sourceCode" id="cb2"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb2-1"><a href="#cb2-1" tabindex="-1"></a>iris.dt[, <span class="fu">fwrite</span>(.SD, <span class="fu">icsv</span>(Species)), by<span class="ot">=</span>Species]</span>
<span id="cb2-2"><a href="#cb2-2" tabindex="-1"></a><span class="co">#&gt; Empty data.table (0 rows and 1 cols): Species</span></span>
<span id="cb2-3"><a href="#cb2-3" tabindex="-1"></a><span class="fu">dir</span>(iris.dir)</span>
<span id="cb2-4"><a href="#cb2-4" tabindex="-1"></a><span class="co">#&gt; [1] &quot;setosa.csv&quot;     &quot;versicolor.csv&quot; &quot;virginica.csv&quot;</span></span></code></pre></div>
<p>The output above shows that there are three CSV files, one for each
Species in the iris data. Below we read the first two rows of one
file,</p>
<div class="sourceCode" id="cb3"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb3-1"><a href="#cb3-1" tabindex="-1"></a>data.table<span class="sc">::</span><span class="fu">fread</span>(<span class="fu">file.path</span>(iris.dir,<span class="st">&quot;setosa.csv&quot;</span>), <span class="at">nrows=</span><span class="dv">2</span>)</span>
<span id="cb3-2"><a href="#cb3-2" tabindex="-1"></a><span class="co">#&gt;    Sepal.Length Sepal.Width Petal.Length Petal.Width</span></span>
<span id="cb3-3"><a href="#cb3-3" tabindex="-1"></a><span class="co">#&gt;           &lt;num&gt;       &lt;num&gt;        &lt;num&gt;       &lt;num&gt;</span></span>
<span id="cb3-4"><a href="#cb3-4" tabindex="-1"></a><span class="co">#&gt; 1:          5.1         3.5          1.4         0.2</span></span>
<span id="cb3-5"><a href="#cb3-5" tabindex="-1"></a><span class="co">#&gt; 2:          4.9         3.0          1.4         0.2</span></span></code></pre></div>
<p>The output above shows that the CSV data file itself does not contain
a Species column (the Species is instead encoded in the file name).
Below we construct a glob, which is a string for matching files,</p>
<div class="sourceCode" id="cb4"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb4-1"><a href="#cb4-1" tabindex="-1"></a>(iglob <span class="ot">&lt;-</span> <span class="fu">file.path</span>(iris.dir,<span class="st">&quot;*.csv&quot;</span>))</span>
<span id="cb4-2"><a href="#cb4-2" tabindex="-1"></a><span class="co">#&gt; [1] &quot;/tmp/RtmpaHnllH/file2d4a3c1e3aef2c/*.csv&quot;</span></span>
<span id="cb4-3"><a href="#cb4-3" tabindex="-1"></a><span class="fu">Sys.glob</span>(iglob)</span>
<span id="cb4-4"><a href="#cb4-4" tabindex="-1"></a><span class="co">#&gt; [1] &quot;/tmp/RtmpaHnllH/file2d4a3c1e3aef2c/setosa.csv&quot;    </span></span>
<span id="cb4-5"><a href="#cb4-5" tabindex="-1"></a><span class="co">#&gt; [2] &quot;/tmp/RtmpaHnllH/file2d4a3c1e3aef2c/versicolor.csv&quot;</span></span>
<span id="cb4-6"><a href="#cb4-6" tabindex="-1"></a><span class="co">#&gt; [3] &quot;/tmp/RtmpaHnllH/file2d4a3c1e3aef2c/virginica.csv&quot;</span></span></code></pre></div>
<p>The output above indicates that <code>iglob</code> matches the three
data files. Below we read those files into R, using the following
syntax:</p>
<ul>
<li>The first argument <code>iglob</code> is a string/glob which
indicates the files to read,</li>
<li>the other arguments form a regular expression pattern:
<ul>
<li>The named argument <code>Species</code> matches that part of the
file name, and is captured to the resulting column of the same
name,</li>
<li>the un-named argument <code>&quot;[.]csv&quot;</code> indicates that suffix
must be matched (but since the argument is not named, it is not
captured, nor saved as a column in the output).</li>
</ul></li>
</ul>
<div class="sourceCode" id="cb5"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb5-1"><a href="#cb5-1" tabindex="-1"></a>nc<span class="sc">::</span><span class="fu">capture_first_glob</span>(iglob, <span class="at">Species=</span><span class="st">&quot;[^/]+&quot;</span>, <span class="st">&quot;[.]csv&quot;</span>)</span>
<span id="cb5-2"><a href="#cb5-2" tabindex="-1"></a><span class="co">#&gt;        Species Sepal.Length Sepal.Width Petal.Length Petal.Width</span></span>
<span id="cb5-3"><a href="#cb5-3" tabindex="-1"></a><span class="co">#&gt;         &lt;char&gt;        &lt;num&gt;       &lt;num&gt;        &lt;num&gt;       &lt;num&gt;</span></span>
<span id="cb5-4"><a href="#cb5-4" tabindex="-1"></a><span class="co">#&gt;   1:    setosa          5.1         3.5          1.4         0.2</span></span>
<span id="cb5-5"><a href="#cb5-5" tabindex="-1"></a><span class="co">#&gt;   2:    setosa          4.9         3.0          1.4         0.2</span></span>
<span id="cb5-6"><a href="#cb5-6" tabindex="-1"></a><span class="co">#&gt;   3:    setosa          4.7         3.2          1.3         0.2</span></span>
<span id="cb5-7"><a href="#cb5-7" tabindex="-1"></a><span class="co">#&gt;   4:    setosa          4.6         3.1          1.5         0.2</span></span>
<span id="cb5-8"><a href="#cb5-8" tabindex="-1"></a><span class="co">#&gt;   5:    setosa          5.0         3.6          1.4         0.2</span></span>
<span id="cb5-9"><a href="#cb5-9" tabindex="-1"></a><span class="co">#&gt;  ---                                                            </span></span>
<span id="cb5-10"><a href="#cb5-10" tabindex="-1"></a><span class="co">#&gt; 146: virginica          6.7         3.0          5.2         2.3</span></span>
<span id="cb5-11"><a href="#cb5-11" tabindex="-1"></a><span class="co">#&gt; 147: virginica          6.3         2.5          5.0         1.9</span></span>
<span id="cb5-12"><a href="#cb5-12" tabindex="-1"></a><span class="co">#&gt; 148: virginica          6.5         3.0          5.2         2.0</span></span>
<span id="cb5-13"><a href="#cb5-13" tabindex="-1"></a><span class="co">#&gt; 149: virginica          6.2         3.4          5.4         2.3</span></span>
<span id="cb5-14"><a href="#cb5-14" tabindex="-1"></a><span class="co">#&gt; 150: virginica          5.9         3.0          5.1         1.8</span></span></code></pre></div>
<p>The output above indicates that we have successfully read the iris
data back into R, including the <code>Species</code> column which was
not present in the CSV data files.</p>
</div>
<div id="example-1-four-files-two-capture-groups-custom-read-function" class="section level2">
<h2>Example 1: four files, two capture groups, custom read function</h2>
<p>Consider the example below, which is slightly more complex. The code
below defines a glob for matching several data files.</p>
<div class="sourceCode" id="cb6"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb6-1"><a href="#cb6-1" tabindex="-1"></a>db <span class="ot">&lt;-</span> <span class="fu">system.file</span>(<span class="st">&quot;extdata/chip-seq-chunk-db&quot;</span>, <span class="at">package=</span><span class="st">&quot;nc&quot;</span>, <span class="at">mustWork=</span><span class="cn">TRUE</span>)</span>
<span id="cb6-2"><a href="#cb6-2" tabindex="-1"></a>(glob <span class="ot">&lt;-</span> <span class="fu">paste0</span>(db, <span class="st">&quot;/*/*/counts/*gz&quot;</span>))</span>
<span id="cb6-3"><a href="#cb6-3" tabindex="-1"></a><span class="co">#&gt; [1] &quot;/tmp/RtmppDHVyz/Rinst2d4985211eda79/nc/extdata/chip-seq-chunk-db/*/*/counts/*gz&quot;</span></span>
<span id="cb6-4"><a href="#cb6-4" tabindex="-1"></a>(matched.files <span class="ot">&lt;-</span> <span class="fu">Sys.glob</span>(glob))</span>
<span id="cb6-5"><a href="#cb6-5" tabindex="-1"></a><span class="co">#&gt; [1] &quot;/tmp/RtmppDHVyz/Rinst2d4985211eda79/nc/extdata/chip-seq-chunk-db/H3K36me3_AM_immune/9/counts/McGill0101.bedGraph.gz&quot;</span></span>
<span id="cb6-6"><a href="#cb6-6" tabindex="-1"></a><span class="co">#&gt; [2] &quot;/tmp/RtmppDHVyz/Rinst2d4985211eda79/nc/extdata/chip-seq-chunk-db/H3K36me3_TDH_other/1/counts/McGill0019.bedGraph.gz&quot;</span></span>
<span id="cb6-7"><a href="#cb6-7" tabindex="-1"></a><span class="co">#&gt; [3] &quot;/tmp/RtmppDHVyz/Rinst2d4985211eda79/nc/extdata/chip-seq-chunk-db/H3K4me3_TDH_immune/9/counts/McGill0024.bedGraph.gz&quot;</span></span>
<span id="cb6-8"><a href="#cb6-8" tabindex="-1"></a><span class="co">#&gt; [4] &quot;/tmp/RtmppDHVyz/Rinst2d4985211eda79/nc/extdata/chip-seq-chunk-db/H3K4me3_XJ_immune/2/counts/McGill0024.bedGraph.gz&quot;</span></span></code></pre></div>
<p>The output above indicates there are four data files that are matched
by the glob. Below we read the first one,</p>
<div class="sourceCode" id="cb7"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb7-1"><a href="#cb7-1" tabindex="-1"></a><span class="fu">readLines</span>(matched.files[<span class="dv">1</span>], <span class="at">n=</span><span class="dv">5</span>)</span>
<span id="cb7-2"><a href="#cb7-2" tabindex="-1"></a><span class="co">#&gt; [1] &quot;track type=bedGraph db=hg19 visibility=full graphType=points name=101K36monocyte description=\&quot;McGill0101 H3K36me3 aligned read counts\&quot;&quot;</span></span>
<span id="cb7-3"><a href="#cb7-3" tabindex="-1"></a><span class="co">#&gt; [2] &quot;chr10\t111456281\t111456338\t2&quot;                                                                                                          </span></span>
<span id="cb7-4"><a href="#cb7-4" tabindex="-1"></a><span class="co">#&gt; [3] &quot;chr10\t111456338\t111456381\t1&quot;                                                                                                          </span></span>
<span id="cb7-5"><a href="#cb7-5" tabindex="-1"></a><span class="co">#&gt; [4] &quot;chr10\t111456381\t111459312\t0&quot;                                                                                                          </span></span>
<span id="cb7-6"><a href="#cb7-6" tabindex="-1"></a><span class="co">#&gt; [5] &quot;chr10\t111459312\t111459316\t5&quot;</span></span></code></pre></div>
<p>We can see from the output above that this data file has a header of
meta-data (not column names) on the first line, whereas the other lines
contain tab-delimited data. We can read it with fread, as long as we
provide a couple non-default arguments, as in the code below:</p>
<div class="sourceCode" id="cb8"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb8-1"><a href="#cb8-1" tabindex="-1"></a>read.bedGraph <span class="ot">&lt;-</span> <span class="cf">function</span>(f)data.table<span class="sc">::</span><span class="fu">fread</span>(</span>
<span id="cb8-2"><a href="#cb8-2" tabindex="-1"></a>  f, <span class="at">skip=</span><span class="dv">1</span>, <span class="at">col.names =</span> <span class="fu">c</span>(<span class="st">&quot;chrom&quot;</span>,<span class="st">&quot;start&quot;</span>, <span class="st">&quot;end&quot;</span>, <span class="st">&quot;count&quot;</span>))</span>
<span id="cb8-3"><a href="#cb8-3" tabindex="-1"></a><span class="fu">read.bedGraph</span>(matched.files[<span class="dv">1</span>])</span>
<span id="cb8-4"><a href="#cb8-4" tabindex="-1"></a><span class="co">#&gt;        chrom     start       end count</span></span>
<span id="cb8-5"><a href="#cb8-5" tabindex="-1"></a><span class="co">#&gt;       &lt;char&gt;     &lt;int&gt;     &lt;int&gt; &lt;int&gt;</span></span>
<span id="cb8-6"><a href="#cb8-6" tabindex="-1"></a><span class="co">#&gt;    1:  chr10 111456281 111456338     2</span></span>
<span id="cb8-7"><a href="#cb8-7" tabindex="-1"></a><span class="co">#&gt;    2:  chr10 111456338 111456381     1</span></span>
<span id="cb8-8"><a href="#cb8-8" tabindex="-1"></a><span class="co">#&gt;    3:  chr10 111456381 111459312     0</span></span>
<span id="cb8-9"><a href="#cb8-9" tabindex="-1"></a><span class="co">#&gt;    4:  chr10 111459312 111459316     5</span></span>
<span id="cb8-10"><a href="#cb8-10" tabindex="-1"></a><span class="co">#&gt;    5:  chr10 111459316 111459409    10</span></span>
<span id="cb8-11"><a href="#cb8-11" tabindex="-1"></a><span class="co">#&gt;   ---                                 </span></span>
<span id="cb8-12"><a href="#cb8-12" tabindex="-1"></a><span class="co">#&gt; 7130:  chr10 111721272 111721347     4</span></span>
<span id="cb8-13"><a href="#cb8-13" tabindex="-1"></a><span class="co">#&gt; 7131:  chr10 111721347 111721354     2</span></span>
<span id="cb8-14"><a href="#cb8-14" tabindex="-1"></a><span class="co">#&gt; 7132:  chr10 111721354 111722459     0</span></span>
<span id="cb8-15"><a href="#cb8-15" tabindex="-1"></a><span class="co">#&gt; 7133:  chr10 111722459 111722461     2</span></span>
<span id="cb8-16"><a href="#cb8-16" tabindex="-1"></a><span class="co">#&gt; 7134:  chr10 111722461 111722555     4</span></span></code></pre></div>
<p>The output above indicates the data has been correctly read into R as
a table with four columns. To do that for each of the files, we use this
custom <code>READ</code> function in the code below,</p>
<div class="sourceCode" id="cb9"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb9-1"><a href="#cb9-1" tabindex="-1"></a>data.chunk.pattern <span class="ot">&lt;-</span> <span class="fu">list</span>(</span>
<span id="cb9-2"><a href="#cb9-2" tabindex="-1"></a>  <span class="at">data=</span><span class="st">&quot;H.*?&quot;</span>,</span>
<span id="cb9-3"><a href="#cb9-3" tabindex="-1"></a>  <span class="st">&quot;/&quot;</span>,</span>
<span id="cb9-4"><a href="#cb9-4" tabindex="-1"></a>  <span class="at">chunk=</span><span class="st">&quot;[0-9]+&quot;</span>, as.integer)</span>
<span id="cb9-5"><a href="#cb9-5" tabindex="-1"></a>(data.chunk.dt <span class="ot">&lt;-</span> nc<span class="sc">::</span><span class="fu">capture_first_glob</span>(glob, data.chunk.pattern, <span class="at">READ=</span>read.bedGraph))</span>
<span id="cb9-6"><a href="#cb9-6" tabindex="-1"></a><span class="co">#&gt;                                                                            data chunk  chrom</span></span>
<span id="cb9-7"><a href="#cb9-7" tabindex="-1"></a><span class="co">#&gt;                                                                          &lt;char&gt; &lt;int&gt; &lt;char&gt;</span></span>
<span id="cb9-8"><a href="#cb9-8" tabindex="-1"></a><span class="co">#&gt;     1: HVyz/Rinst2d4985211eda79/nc/extdata/chip-seq-chunk-db/H3K36me3_AM_immune     9  chr10</span></span>
<span id="cb9-9"><a href="#cb9-9" tabindex="-1"></a><span class="co">#&gt;     2: HVyz/Rinst2d4985211eda79/nc/extdata/chip-seq-chunk-db/H3K36me3_AM_immune     9  chr10</span></span>
<span id="cb9-10"><a href="#cb9-10" tabindex="-1"></a><span class="co">#&gt;     3: HVyz/Rinst2d4985211eda79/nc/extdata/chip-seq-chunk-db/H3K36me3_AM_immune     9  chr10</span></span>
<span id="cb9-11"><a href="#cb9-11" tabindex="-1"></a><span class="co">#&gt;     4: HVyz/Rinst2d4985211eda79/nc/extdata/chip-seq-chunk-db/H3K36me3_AM_immune     9  chr10</span></span>
<span id="cb9-12"><a href="#cb9-12" tabindex="-1"></a><span class="co">#&gt;     5: HVyz/Rinst2d4985211eda79/nc/extdata/chip-seq-chunk-db/H3K36me3_AM_immune     9  chr10</span></span>
<span id="cb9-13"><a href="#cb9-13" tabindex="-1"></a><span class="co">#&gt;    ---                                                                                      </span></span>
<span id="cb9-14"><a href="#cb9-14" tabindex="-1"></a><span class="co">#&gt; 20297:  HVyz/Rinst2d4985211eda79/nc/extdata/chip-seq-chunk-db/H3K4me3_XJ_immune     2  chr22</span></span>
<span id="cb9-15"><a href="#cb9-15" tabindex="-1"></a><span class="co">#&gt; 20298:  HVyz/Rinst2d4985211eda79/nc/extdata/chip-seq-chunk-db/H3K4me3_XJ_immune     2  chr22</span></span>
<span id="cb9-16"><a href="#cb9-16" tabindex="-1"></a><span class="co">#&gt; 20299:  HVyz/Rinst2d4985211eda79/nc/extdata/chip-seq-chunk-db/H3K4me3_XJ_immune     2  chr22</span></span>
<span id="cb9-17"><a href="#cb9-17" tabindex="-1"></a><span class="co">#&gt; 20300:  HVyz/Rinst2d4985211eda79/nc/extdata/chip-seq-chunk-db/H3K4me3_XJ_immune     2  chr22</span></span>
<span id="cb9-18"><a href="#cb9-18" tabindex="-1"></a><span class="co">#&gt; 20301:  HVyz/Rinst2d4985211eda79/nc/extdata/chip-seq-chunk-db/H3K4me3_XJ_immune     2  chr22</span></span>
<span id="cb9-19"><a href="#cb9-19" tabindex="-1"></a><span class="co">#&gt;            start       end count</span></span>
<span id="cb9-20"><a href="#cb9-20" tabindex="-1"></a><span class="co">#&gt;            &lt;int&gt;     &lt;int&gt; &lt;int&gt;</span></span>
<span id="cb9-21"><a href="#cb9-21" tabindex="-1"></a><span class="co">#&gt;     1: 111456281 111456338     2</span></span>
<span id="cb9-22"><a href="#cb9-22" tabindex="-1"></a><span class="co">#&gt;     2: 111456338 111456381     1</span></span>
<span id="cb9-23"><a href="#cb9-23" tabindex="-1"></a><span class="co">#&gt;     3: 111456381 111459312     0</span></span>
<span id="cb9-24"><a href="#cb9-24" tabindex="-1"></a><span class="co">#&gt;     4: 111459312 111459316     5</span></span>
<span id="cb9-25"><a href="#cb9-25" tabindex="-1"></a><span class="co">#&gt;     5: 111459316 111459409    10</span></span>
<span id="cb9-26"><a href="#cb9-26" tabindex="-1"></a><span class="co">#&gt;    ---                          </span></span>
<span id="cb9-27"><a href="#cb9-27" tabindex="-1"></a><span class="co">#&gt; 20297:  20689768  20689770     0</span></span>
<span id="cb9-28"><a href="#cb9-28" tabindex="-1"></a><span class="co">#&gt; 20298:  20689770  20689870     1</span></span>
<span id="cb9-29"><a href="#cb9-29" tabindex="-1"></a><span class="co">#&gt; 20299:  20689870  20689995     0</span></span>
<span id="cb9-30"><a href="#cb9-30" tabindex="-1"></a><span class="co">#&gt; 20300:  20689995  20690080     1</span></span>
<span id="cb9-31"><a href="#cb9-31" tabindex="-1"></a><span class="co">#&gt; 20301:  20690080  20691400     0</span></span></code></pre></div>
<p>The output above indicates the data files have been read into R as a
table, with two additional columns (data and chunk), which correspond to
the capture group names used in the regular expression pattern
above.</p>
</div>
<div id="why-not-base-r" class="section level2">
<h2>Why not base R?</h2>
<p>We can absolutely use base R to read these files, but it takes a bit
more code, as shown below.</p>
<div class="sourceCode" id="cb10"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb10-1"><a href="#cb10-1" tabindex="-1"></a>base.df.list <span class="ot">&lt;-</span> <span class="fu">list</span>()</span>
<span id="cb10-2"><a href="#cb10-2" tabindex="-1"></a><span class="cf">for</span>(file.csv <span class="cf">in</span> matched.files){</span>
<span id="cb10-3"><a href="#cb10-3" tabindex="-1"></a>  file.df <span class="ot">&lt;-</span> <span class="fu">read.bedGraph</span>(file.csv)</span>
<span id="cb10-4"><a href="#cb10-4" tabindex="-1"></a>  counts.path <span class="ot">&lt;-</span> <span class="fu">dirname</span>(file.csv)</span>
<span id="cb10-5"><a href="#cb10-5" tabindex="-1"></a>  chunk.path <span class="ot">&lt;-</span> <span class="fu">dirname</span>(counts.path)</span>
<span id="cb10-6"><a href="#cb10-6" tabindex="-1"></a>  data.path <span class="ot">&lt;-</span> <span class="fu">dirname</span>(chunk.path)</span>
<span id="cb10-7"><a href="#cb10-7" tabindex="-1"></a>  base.df.list[[file.csv]] <span class="ot">&lt;-</span> <span class="fu">data.frame</span>(</span>
<span id="cb10-8"><a href="#cb10-8" tabindex="-1"></a>    <span class="at">data=</span><span class="fu">basename</span>(data.path),</span>
<span id="cb10-9"><a href="#cb10-9" tabindex="-1"></a>    <span class="at">chunk=</span><span class="fu">basename</span>(chunk.path),</span>
<span id="cb10-10"><a href="#cb10-10" tabindex="-1"></a>    file.df)</span>
<span id="cb10-11"><a href="#cb10-11" tabindex="-1"></a>}</span>
<span id="cb10-12"><a href="#cb10-12" tabindex="-1"></a>base.df <span class="ot">&lt;-</span> <span class="fu">do.call</span>(rbind, base.df.list)</span>
<span id="cb10-13"><a href="#cb10-13" tabindex="-1"></a><span class="fu">rownames</span>(base.df) <span class="ot">&lt;-</span> <span class="cn">NULL</span></span>
<span id="cb10-14"><a href="#cb10-14" tabindex="-1"></a><span class="fu">head</span>(base.df)</span>
<span id="cb10-15"><a href="#cb10-15" tabindex="-1"></a><span class="co">#&gt;                 data chunk chrom     start       end count</span></span>
<span id="cb10-16"><a href="#cb10-16" tabindex="-1"></a><span class="co">#&gt; 1 H3K36me3_AM_immune     9 chr10 111456281 111456338     2</span></span>
<span id="cb10-17"><a href="#cb10-17" tabindex="-1"></a><span class="co">#&gt; 2 H3K36me3_AM_immune     9 chr10 111456338 111456381     1</span></span>
<span id="cb10-18"><a href="#cb10-18" tabindex="-1"></a><span class="co">#&gt; 3 H3K36me3_AM_immune     9 chr10 111456381 111459312     0</span></span>
<span id="cb10-19"><a href="#cb10-19" tabindex="-1"></a><span class="co">#&gt; 4 H3K36me3_AM_immune     9 chr10 111459312 111459316     5</span></span>
<span id="cb10-20"><a href="#cb10-20" tabindex="-1"></a><span class="co">#&gt; 5 H3K36me3_AM_immune     9 chr10 111459316 111459409    10</span></span>
<span id="cb10-21"><a href="#cb10-21" tabindex="-1"></a><span class="co">#&gt; 6 H3K36me3_AM_immune     9 chr10 111459409 111459411     8</span></span>
<span id="cb10-22"><a href="#cb10-22" tabindex="-1"></a><span class="fu">str</span>(base.df)</span>
<span id="cb10-23"><a href="#cb10-23" tabindex="-1"></a><span class="co">#&gt; &#39;data.frame&#39;:    20301 obs. of  6 variables:</span></span>
<span id="cb10-24"><a href="#cb10-24" tabindex="-1"></a><span class="co">#&gt;  $ data : chr  &quot;H3K36me3_AM_immune&quot; &quot;H3K36me3_AM_immune&quot; &quot;H3K36me3_AM_immune&quot; &quot;H3K36me3_AM_immune&quot; ...</span></span>
<span id="cb10-25"><a href="#cb10-25" tabindex="-1"></a><span class="co">#&gt;  $ chunk: chr  &quot;9&quot; &quot;9&quot; &quot;9&quot; &quot;9&quot; ...</span></span>
<span id="cb10-26"><a href="#cb10-26" tabindex="-1"></a><span class="co">#&gt;  $ chrom: chr  &quot;chr10&quot; &quot;chr10&quot; &quot;chr10&quot; &quot;chr10&quot; ...</span></span>
<span id="cb10-27"><a href="#cb10-27" tabindex="-1"></a><span class="co">#&gt;  $ start: int  111456281 111456338 111456381 111459312 111459316 111459409 111459411 111459415 111463412 111463512 ...</span></span>
<span id="cb10-28"><a href="#cb10-28" tabindex="-1"></a><span class="co">#&gt;  $ end  : int  111456338 111456381 111459312 111459316 111459409 111459411 111459415 111463412 111463512 111466726 ...</span></span>
<span id="cb10-29"><a href="#cb10-29" tabindex="-1"></a><span class="co">#&gt;  $ count: int  2 1 0 5 10 8 5 0 2 0 ...</span></span></code></pre></div>
<p>The output above shows that we have read a data frame into R, and
that it is consistent with the data table returned by
<code>nc::capture_first_glob</code>, which should be preferred for
simplicity when the files are regularly named. In contrast, this section
shows how arbitrary R code can be used, so this approach should be
preferred when the data in the file path can not be captured using
regular expressions.</p>
</div>
<div id="example-3-hive-partition-file-names" class="section level2">
<h2>Example 3: Hive partition file names</h2>
<p>In the code below, we write the same data to a set of CSV files with
different names,</p>
<div class="sourceCode" id="cb11"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb11-1"><a href="#cb11-1" tabindex="-1"></a>arrow.available <span class="ot">&lt;-</span> <span class="fu">requireNamespace</span>(<span class="st">&quot;arrow&quot;</span>) <span class="sc">&amp;&amp;</span> arrow<span class="sc">::</span><span class="fu">arrow_with_dataset</span>()</span>
<span id="cb11-2"><a href="#cb11-2" tabindex="-1"></a><span class="co">#&gt; Le chargement a n?cessit? le package : arrow</span></span>
<span id="cb11-3"><a href="#cb11-3" tabindex="-1"></a><span class="cf">if</span>(arrow.available){</span>
<span id="cb11-4"><a href="#cb11-4" tabindex="-1"></a>  path <span class="ot">&lt;-</span> <span class="fu">tempfile</span>()</span>
<span id="cb11-5"><a href="#cb11-5" tabindex="-1"></a>  arrow<span class="sc">::</span><span class="fu">write_dataset</span>(</span>
<span id="cb11-6"><a href="#cb11-6" tabindex="-1"></a>    <span class="at">dataset=</span>data.chunk.dt,</span>
<span id="cb11-7"><a href="#cb11-7" tabindex="-1"></a>    <span class="at">path=</span>path,</span>
<span id="cb11-8"><a href="#cb11-8" tabindex="-1"></a>    <span class="at">format=</span><span class="st">&quot;csv&quot;</span>,</span>
<span id="cb11-9"><a href="#cb11-9" tabindex="-1"></a>    <span class="at">partitioning=</span><span class="fu">c</span>(<span class="st">&quot;data&quot;</span>,<span class="st">&quot;chunk&quot;</span>),</span>
<span id="cb11-10"><a href="#cb11-10" tabindex="-1"></a>    <span class="at">max_rows_per_file=</span><span class="dv">1000</span>)</span>
<span id="cb11-11"><a href="#cb11-11" tabindex="-1"></a>  hive.glob <span class="ot">&lt;-</span> <span class="fu">file.path</span>(path, <span class="st">&quot;*&quot;</span>, <span class="st">&quot;*&quot;</span>, <span class="st">&quot;*.csv&quot;</span>)</span>
<span id="cb11-12"><a href="#cb11-12" tabindex="-1"></a>  (hive.files <span class="ot">&lt;-</span> <span class="fu">Sys.glob</span>(hive.glob))</span>
<span id="cb11-13"><a href="#cb11-13" tabindex="-1"></a>}</span>
<span id="cb11-14"><a href="#cb11-14" tabindex="-1"></a><span class="co">#&gt;  [1] &quot;/tmp/RtmpaHnllH/file2d4a3c57bdf6fb/data=HVyz%2FRinst2d4985211eda79%2Fnc%2Fextdata%2Fchip-seq-chunk-db%2FH3K36me3_AM_immune/chunk=9/part-0.csv&quot; </span></span>
<span id="cb11-15"><a href="#cb11-15" tabindex="-1"></a><span class="co">#&gt;  [2] &quot;/tmp/RtmpaHnllH/file2d4a3c57bdf6fb/data=HVyz%2FRinst2d4985211eda79%2Fnc%2Fextdata%2Fchip-seq-chunk-db%2FH3K36me3_AM_immune/chunk=9/part-1.csv&quot; </span></span>
<span id="cb11-16"><a href="#cb11-16" tabindex="-1"></a><span class="co">#&gt;  [3] &quot;/tmp/RtmpaHnllH/file2d4a3c57bdf6fb/data=HVyz%2FRinst2d4985211eda79%2Fnc%2Fextdata%2Fchip-seq-chunk-db%2FH3K36me3_AM_immune/chunk=9/part-2.csv&quot; </span></span>
<span id="cb11-17"><a href="#cb11-17" tabindex="-1"></a><span class="co">#&gt;  [4] &quot;/tmp/RtmpaHnllH/file2d4a3c57bdf6fb/data=HVyz%2FRinst2d4985211eda79%2Fnc%2Fextdata%2Fchip-seq-chunk-db%2FH3K36me3_AM_immune/chunk=9/part-3.csv&quot; </span></span>
<span id="cb11-18"><a href="#cb11-18" tabindex="-1"></a><span class="co">#&gt;  [5] &quot;/tmp/RtmpaHnllH/file2d4a3c57bdf6fb/data=HVyz%2FRinst2d4985211eda79%2Fnc%2Fextdata%2Fchip-seq-chunk-db%2FH3K36me3_AM_immune/chunk=9/part-4.csv&quot; </span></span>
<span id="cb11-19"><a href="#cb11-19" tabindex="-1"></a><span class="co">#&gt;  [6] &quot;/tmp/RtmpaHnllH/file2d4a3c57bdf6fb/data=HVyz%2FRinst2d4985211eda79%2Fnc%2Fextdata%2Fchip-seq-chunk-db%2FH3K36me3_AM_immune/chunk=9/part-5.csv&quot; </span></span>
<span id="cb11-20"><a href="#cb11-20" tabindex="-1"></a><span class="co">#&gt;  [7] &quot;/tmp/RtmpaHnllH/file2d4a3c57bdf6fb/data=HVyz%2FRinst2d4985211eda79%2Fnc%2Fextdata%2Fchip-seq-chunk-db%2FH3K36me3_AM_immune/chunk=9/part-6.csv&quot; </span></span>
<span id="cb11-21"><a href="#cb11-21" tabindex="-1"></a><span class="co">#&gt;  [8] &quot;/tmp/RtmpaHnllH/file2d4a3c57bdf6fb/data=HVyz%2FRinst2d4985211eda79%2Fnc%2Fextdata%2Fchip-seq-chunk-db%2FH3K36me3_AM_immune/chunk=9/part-7.csv&quot; </span></span>
<span id="cb11-22"><a href="#cb11-22" tabindex="-1"></a><span class="co">#&gt;  [9] &quot;/tmp/RtmpaHnllH/file2d4a3c57bdf6fb/data=HVyz%2FRinst2d4985211eda79%2Fnc%2Fextdata%2Fchip-seq-chunk-db%2FH3K36me3_TDH_other/chunk=1/part-0.csv&quot; </span></span>
<span id="cb11-23"><a href="#cb11-23" tabindex="-1"></a><span class="co">#&gt; [10] &quot;/tmp/RtmpaHnllH/file2d4a3c57bdf6fb/data=HVyz%2FRinst2d4985211eda79%2Fnc%2Fextdata%2Fchip-seq-chunk-db%2FH3K36me3_TDH_other/chunk=1/part-1.csv&quot; </span></span>
<span id="cb11-24"><a href="#cb11-24" tabindex="-1"></a><span class="co">#&gt; [11] &quot;/tmp/RtmpaHnllH/file2d4a3c57bdf6fb/data=HVyz%2FRinst2d4985211eda79%2Fnc%2Fextdata%2Fchip-seq-chunk-db%2FH3K36me3_TDH_other/chunk=1/part-10.csv&quot;</span></span>
<span id="cb11-25"><a href="#cb11-25" tabindex="-1"></a><span class="co">#&gt; [12] &quot;/tmp/RtmpaHnllH/file2d4a3c57bdf6fb/data=HVyz%2FRinst2d4985211eda79%2Fnc%2Fextdata%2Fchip-seq-chunk-db%2FH3K36me3_TDH_other/chunk=1/part-11.csv&quot;</span></span>
<span id="cb11-26"><a href="#cb11-26" tabindex="-1"></a><span class="co">#&gt; [13] &quot;/tmp/RtmpaHnllH/file2d4a3c57bdf6fb/data=HVyz%2FRinst2d4985211eda79%2Fnc%2Fextdata%2Fchip-seq-chunk-db%2FH3K36me3_TDH_other/chunk=1/part-12.csv&quot;</span></span>
<span id="cb11-27"><a href="#cb11-27" tabindex="-1"></a><span class="co">#&gt; [14] &quot;/tmp/RtmpaHnllH/file2d4a3c57bdf6fb/data=HVyz%2FRinst2d4985211eda79%2Fnc%2Fextdata%2Fchip-seq-chunk-db%2FH3K36me3_TDH_other/chunk=1/part-2.csv&quot; </span></span>
<span id="cb11-28"><a href="#cb11-28" tabindex="-1"></a><span class="co">#&gt; [15] &quot;/tmp/RtmpaHnllH/file2d4a3c57bdf6fb/data=HVyz%2FRinst2d4985211eda79%2Fnc%2Fextdata%2Fchip-seq-chunk-db%2FH3K36me3_TDH_other/chunk=1/part-3.csv&quot; </span></span>
<span id="cb11-29"><a href="#cb11-29" tabindex="-1"></a><span class="co">#&gt; [16] &quot;/tmp/RtmpaHnllH/file2d4a3c57bdf6fb/data=HVyz%2FRinst2d4985211eda79%2Fnc%2Fextdata%2Fchip-seq-chunk-db%2FH3K36me3_TDH_other/chunk=1/part-4.csv&quot; </span></span>
<span id="cb11-30"><a href="#cb11-30" tabindex="-1"></a><span class="co">#&gt; [17] &quot;/tmp/RtmpaHnllH/file2d4a3c57bdf6fb/data=HVyz%2FRinst2d4985211eda79%2Fnc%2Fextdata%2Fchip-seq-chunk-db%2FH3K36me3_TDH_other/chunk=1/part-5.csv&quot; </span></span>
<span id="cb11-31"><a href="#cb11-31" tabindex="-1"></a><span class="co">#&gt; [18] &quot;/tmp/RtmpaHnllH/file2d4a3c57bdf6fb/data=HVyz%2FRinst2d4985211eda79%2Fnc%2Fextdata%2Fchip-seq-chunk-db%2FH3K36me3_TDH_other/chunk=1/part-6.csv&quot; </span></span>
<span id="cb11-32"><a href="#cb11-32" tabindex="-1"></a><span class="co">#&gt; [19] &quot;/tmp/RtmpaHnllH/file2d4a3c57bdf6fb/data=HVyz%2FRinst2d4985211eda79%2Fnc%2Fextdata%2Fchip-seq-chunk-db%2FH3K36me3_TDH_other/chunk=1/part-7.csv&quot; </span></span>
<span id="cb11-33"><a href="#cb11-33" tabindex="-1"></a><span class="co">#&gt; [20] &quot;/tmp/RtmpaHnllH/file2d4a3c57bdf6fb/data=HVyz%2FRinst2d4985211eda79%2Fnc%2Fextdata%2Fchip-seq-chunk-db%2FH3K36me3_TDH_other/chunk=1/part-8.csv&quot; </span></span>
<span id="cb11-34"><a href="#cb11-34" tabindex="-1"></a><span class="co">#&gt; [21] &quot;/tmp/RtmpaHnllH/file2d4a3c57bdf6fb/data=HVyz%2FRinst2d4985211eda79%2Fnc%2Fextdata%2Fchip-seq-chunk-db%2FH3K36me3_TDH_other/chunk=1/part-9.csv&quot; </span></span>
<span id="cb11-35"><a href="#cb11-35" tabindex="-1"></a><span class="co">#&gt; [22] &quot;/tmp/RtmpaHnllH/file2d4a3c57bdf6fb/data=HVyz%2FRinst2d4985211eda79%2Fnc%2Fextdata%2Fchip-seq-chunk-db%2FH3K4me3_TDH_immune/chunk=9/part-0.csv&quot; </span></span>
<span id="cb11-36"><a href="#cb11-36" tabindex="-1"></a><span class="co">#&gt; [23] &quot;/tmp/RtmpaHnllH/file2d4a3c57bdf6fb/data=HVyz%2FRinst2d4985211eda79%2Fnc%2Fextdata%2Fchip-seq-chunk-db%2FH3K4me3_XJ_immune/chunk=2/part-0.csv&quot;</span></span></code></pre></div>
<p>In the output above, we can see that there are regularly named files
with three variables encoded in the file path (data, chunk, part). The
code below reads one of the files back into R:</p>
<div class="sourceCode" id="cb12"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb12-1"><a href="#cb12-1" tabindex="-1"></a><span class="cf">if</span>(arrow.available){</span>
<span id="cb12-2"><a href="#cb12-2" tabindex="-1"></a>  data.table<span class="sc">::</span><span class="fu">fread</span>(hive.files[<span class="dv">1</span>])</span>
<span id="cb12-3"><a href="#cb12-3" tabindex="-1"></a>}</span>
<span id="cb12-4"><a href="#cb12-4" tabindex="-1"></a><span class="co">#&gt;        chrom     start       end count</span></span>
<span id="cb12-5"><a href="#cb12-5" tabindex="-1"></a><span class="co">#&gt;       &lt;char&gt;     &lt;int&gt;     &lt;int&gt; &lt;int&gt;</span></span>
<span id="cb12-6"><a href="#cb12-6" tabindex="-1"></a><span class="co">#&gt;    1:  chr10 111456281 111456338     2</span></span>
<span id="cb12-7"><a href="#cb12-7" tabindex="-1"></a><span class="co">#&gt;    2:  chr10 111456338 111456381     1</span></span>
<span id="cb12-8"><a href="#cb12-8" tabindex="-1"></a><span class="co">#&gt;    3:  chr10 111456381 111459312     0</span></span>
<span id="cb12-9"><a href="#cb12-9" tabindex="-1"></a><span class="co">#&gt;    4:  chr10 111459312 111459316     5</span></span>
<span id="cb12-10"><a href="#cb12-10" tabindex="-1"></a><span class="co">#&gt;    5:  chr10 111459316 111459409    10</span></span>
<span id="cb12-11"><a href="#cb12-11" tabindex="-1"></a><span class="co">#&gt;   ---                                 </span></span>
<span id="cb12-12"><a href="#cb12-12" tabindex="-1"></a><span class="co">#&gt;  996:  chr10 111619010 111619035     1</span></span>
<span id="cb12-13"><a href="#cb12-13" tabindex="-1"></a><span class="co">#&gt;  997:  chr10 111619035 111619092     2</span></span>
<span id="cb12-14"><a href="#cb12-14" tabindex="-1"></a><span class="co">#&gt;  998:  chr10 111619092 111619101     3</span></span>
<span id="cb12-15"><a href="#cb12-15" tabindex="-1"></a><span class="co">#&gt;  999:  chr10 111619101 111619128     2</span></span>
<span id="cb12-16"><a href="#cb12-16" tabindex="-1"></a><span class="co">#&gt; 1000:  chr10 111619128 111619129     3</span></span></code></pre></div>
<p>The output above indicates that the file only has four columns (and
is missing the variables which are encoded in the file path). In the
code below, we read all those files back into R:</p>
<div class="sourceCode" id="cb13"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb13-1"><a href="#cb13-1" tabindex="-1"></a><span class="cf">if</span>(arrow.available){</span>
<span id="cb13-2"><a href="#cb13-2" tabindex="-1"></a>  hive.pattern <span class="ot">&lt;-</span> <span class="fu">list</span>(</span>
<span id="cb13-3"><a href="#cb13-3" tabindex="-1"></a>    nc<span class="sc">::</span><span class="fu">field</span>(<span class="st">&quot;data&quot;</span>,<span class="st">&quot;=&quot;</span>,<span class="st">&quot;.*?&quot;</span>),</span>
<span id="cb13-4"><a href="#cb13-4" tabindex="-1"></a>    <span class="st">&quot;/&quot;</span>,</span>
<span id="cb13-5"><a href="#cb13-5" tabindex="-1"></a>    nc<span class="sc">::</span><span class="fu">field</span>(<span class="st">&quot;chunk&quot;</span>,<span class="st">&quot;=&quot;</span>,<span class="st">&quot;.*?&quot;</span>, as.integer),</span>
<span id="cb13-6"><a href="#cb13-6" tabindex="-1"></a>    <span class="st">&quot;/&quot;</span>,</span>
<span id="cb13-7"><a href="#cb13-7" tabindex="-1"></a>    nc<span class="sc">::</span><span class="fu">field</span>(<span class="st">&quot;part&quot;</span>,<span class="st">&quot;-&quot;</span>,<span class="st">&quot;[0-9]+&quot;</span>, as.integer))</span>
<span id="cb13-8"><a href="#cb13-8" tabindex="-1"></a>  <span class="fu">print</span>(hive.dt <span class="ot">&lt;-</span> nc<span class="sc">::</span><span class="fu">capture_first_glob</span>(hive.glob, hive.pattern))</span>
<span id="cb13-9"><a href="#cb13-9" tabindex="-1"></a>  hive.dt[, .(<span class="at">rows=</span>.N), keyby<span class="ot">=</span>.(data,chunk,part)]</span>
<span id="cb13-10"><a href="#cb13-10" tabindex="-1"></a>}</span>
<span id="cb13-11"><a href="#cb13-11" tabindex="-1"></a><span class="co">#&gt;                                                                                      data chunk</span></span>
<span id="cb13-12"><a href="#cb13-12" tabindex="-1"></a><span class="co">#&gt;                                                                                    &lt;char&gt; &lt;int&gt;</span></span>
<span id="cb13-13"><a href="#cb13-13" tabindex="-1"></a><span class="co">#&gt;     1: HVyz%2FRinst2d4985211eda79%2Fnc%2Fextdata%2Fchip-seq-chunk-db%2FH3K36me3_AM_immune     9</span></span>
<span id="cb13-14"><a href="#cb13-14" tabindex="-1"></a><span class="co">#&gt;     2: HVyz%2FRinst2d4985211eda79%2Fnc%2Fextdata%2Fchip-seq-chunk-db%2FH3K36me3_AM_immune     9</span></span>
<span id="cb13-15"><a href="#cb13-15" tabindex="-1"></a><span class="co">#&gt;     3: HVyz%2FRinst2d4985211eda79%2Fnc%2Fextdata%2Fchip-seq-chunk-db%2FH3K36me3_AM_immune     9</span></span>
<span id="cb13-16"><a href="#cb13-16" tabindex="-1"></a><span class="co">#&gt;     4: HVyz%2FRinst2d4985211eda79%2Fnc%2Fextdata%2Fchip-seq-chunk-db%2FH3K36me3_AM_immune     9</span></span>
<span id="cb13-17"><a href="#cb13-17" tabindex="-1"></a><span class="co">#&gt;     5: HVyz%2FRinst2d4985211eda79%2Fnc%2Fextdata%2Fchip-seq-chunk-db%2FH3K36me3_AM_immune     9</span></span>
<span id="cb13-18"><a href="#cb13-18" tabindex="-1"></a><span class="co">#&gt;    ---                                                                                         </span></span>
<span id="cb13-19"><a href="#cb13-19" tabindex="-1"></a><span class="co">#&gt; 20297:  HVyz%2FRinst2d4985211eda79%2Fnc%2Fextdata%2Fchip-seq-chunk-db%2FH3K4me3_XJ_immune     2</span></span>
<span id="cb13-20"><a href="#cb13-20" tabindex="-1"></a><span class="co">#&gt; 20298:  HVyz%2FRinst2d4985211eda79%2Fnc%2Fextdata%2Fchip-seq-chunk-db%2FH3K4me3_XJ_immune     2</span></span>
<span id="cb13-21"><a href="#cb13-21" tabindex="-1"></a><span class="co">#&gt; 20299:  HVyz%2FRinst2d4985211eda79%2Fnc%2Fextdata%2Fchip-seq-chunk-db%2FH3K4me3_XJ_immune     2</span></span>
<span id="cb13-22"><a href="#cb13-22" tabindex="-1"></a><span class="co">#&gt; 20300:  HVyz%2FRinst2d4985211eda79%2Fnc%2Fextdata%2Fchip-seq-chunk-db%2FH3K4me3_XJ_immune     2</span></span>
<span id="cb13-23"><a href="#cb13-23" tabindex="-1"></a><span class="co">#&gt; 20301:  HVyz%2FRinst2d4985211eda79%2Fnc%2Fextdata%2Fchip-seq-chunk-db%2FH3K4me3_XJ_immune     2</span></span>
<span id="cb13-24"><a href="#cb13-24" tabindex="-1"></a><span class="co">#&gt;         part  chrom     start       end count</span></span>
<span id="cb13-25"><a href="#cb13-25" tabindex="-1"></a><span class="co">#&gt;        &lt;int&gt; &lt;char&gt;     &lt;int&gt;     &lt;int&gt; &lt;int&gt;</span></span>
<span id="cb13-26"><a href="#cb13-26" tabindex="-1"></a><span class="co">#&gt;     1:     0  chr10 111456281 111456338     2</span></span>
<span id="cb13-27"><a href="#cb13-27" tabindex="-1"></a><span class="co">#&gt;     2:     0  chr10 111456338 111456381     1</span></span>
<span id="cb13-28"><a href="#cb13-28" tabindex="-1"></a><span class="co">#&gt;     3:     0  chr10 111456381 111459312     0</span></span>
<span id="cb13-29"><a href="#cb13-29" tabindex="-1"></a><span class="co">#&gt;     4:     0  chr10 111459312 111459316     5</span></span>
<span id="cb13-30"><a href="#cb13-30" tabindex="-1"></a><span class="co">#&gt;     5:     0  chr10 111459316 111459409    10</span></span>
<span id="cb13-31"><a href="#cb13-31" tabindex="-1"></a><span class="co">#&gt;    ---                                       </span></span>
<span id="cb13-32"><a href="#cb13-32" tabindex="-1"></a><span class="co">#&gt; 20297:     0  chr22  20689768  20689770     0</span></span>
<span id="cb13-33"><a href="#cb13-33" tabindex="-1"></a><span class="co">#&gt; 20298:     0  chr22  20689770  20689870     1</span></span>
<span id="cb13-34"><a href="#cb13-34" tabindex="-1"></a><span class="co">#&gt; 20299:     0  chr22  20689870  20689995     0</span></span>
<span id="cb13-35"><a href="#cb13-35" tabindex="-1"></a><span class="co">#&gt; 20300:     0  chr22  20689995  20690080     1</span></span>
<span id="cb13-36"><a href="#cb13-36" tabindex="-1"></a><span class="co">#&gt; 20301:     0  chr22  20690080  20691400     0</span></span>
<span id="cb13-37"><a href="#cb13-37" tabindex="-1"></a><span class="co">#&gt; Key: &lt;data, chunk, part&gt;</span></span>
<span id="cb13-38"><a href="#cb13-38" tabindex="-1"></a><span class="co">#&gt;                                                                                   data chunk  part</span></span>
<span id="cb13-39"><a href="#cb13-39" tabindex="-1"></a><span class="co">#&gt;                                                                                 &lt;char&gt; &lt;int&gt; &lt;int&gt;</span></span>
<span id="cb13-40"><a href="#cb13-40" tabindex="-1"></a><span class="co">#&gt;  1: HVyz%2FRinst2d4985211eda79%2Fnc%2Fextdata%2Fchip-seq-chunk-db%2FH3K36me3_AM_immune     9     0</span></span>
<span id="cb13-41"><a href="#cb13-41" tabindex="-1"></a><span class="co">#&gt;  2: HVyz%2FRinst2d4985211eda79%2Fnc%2Fextdata%2Fchip-seq-chunk-db%2FH3K36me3_AM_immune     9     1</span></span>
<span id="cb13-42"><a href="#cb13-42" tabindex="-1"></a><span class="co">#&gt;  3: HVyz%2FRinst2d4985211eda79%2Fnc%2Fextdata%2Fchip-seq-chunk-db%2FH3K36me3_AM_immune     9     2</span></span>
<span id="cb13-43"><a href="#cb13-43" tabindex="-1"></a><span class="co">#&gt;  4: HVyz%2FRinst2d4985211eda79%2Fnc%2Fextdata%2Fchip-seq-chunk-db%2FH3K36me3_AM_immune     9     3</span></span>
<span id="cb13-44"><a href="#cb13-44" tabindex="-1"></a><span class="co">#&gt;  5: HVyz%2FRinst2d4985211eda79%2Fnc%2Fextdata%2Fchip-seq-chunk-db%2FH3K36me3_AM_immune     9     4</span></span>
<span id="cb13-45"><a href="#cb13-45" tabindex="-1"></a><span class="co">#&gt;  6: HVyz%2FRinst2d4985211eda79%2Fnc%2Fextdata%2Fchip-seq-chunk-db%2FH3K36me3_AM_immune     9     5</span></span>
<span id="cb13-46"><a href="#cb13-46" tabindex="-1"></a><span class="co">#&gt;  7: HVyz%2FRinst2d4985211eda79%2Fnc%2Fextdata%2Fchip-seq-chunk-db%2FH3K36me3_AM_immune     9     6</span></span>
<span id="cb13-47"><a href="#cb13-47" tabindex="-1"></a><span class="co">#&gt;  8: HVyz%2FRinst2d4985211eda79%2Fnc%2Fextdata%2Fchip-seq-chunk-db%2FH3K36me3_AM_immune     9     7</span></span>
<span id="cb13-48"><a href="#cb13-48" tabindex="-1"></a><span class="co">#&gt;  9: HVyz%2FRinst2d4985211eda79%2Fnc%2Fextdata%2Fchip-seq-chunk-db%2FH3K36me3_TDH_other     1     0</span></span>
<span id="cb13-49"><a href="#cb13-49" tabindex="-1"></a><span class="co">#&gt; 10: HVyz%2FRinst2d4985211eda79%2Fnc%2Fextdata%2Fchip-seq-chunk-db%2FH3K36me3_TDH_other     1     1</span></span>
<span id="cb13-50"><a href="#cb13-50" tabindex="-1"></a><span class="co">#&gt; 11: HVyz%2FRinst2d4985211eda79%2Fnc%2Fextdata%2Fchip-seq-chunk-db%2FH3K36me3_TDH_other     1     2</span></span>
<span id="cb13-51"><a href="#cb13-51" tabindex="-1"></a><span class="co">#&gt; 12: HVyz%2FRinst2d4985211eda79%2Fnc%2Fextdata%2Fchip-seq-chunk-db%2FH3K36me3_TDH_other     1     3</span></span>
<span id="cb13-52"><a href="#cb13-52" tabindex="-1"></a><span class="co">#&gt; 13: HVyz%2FRinst2d4985211eda79%2Fnc%2Fextdata%2Fchip-seq-chunk-db%2FH3K36me3_TDH_other     1     4</span></span>
<span id="cb13-53"><a href="#cb13-53" tabindex="-1"></a><span class="co">#&gt; 14: HVyz%2FRinst2d4985211eda79%2Fnc%2Fextdata%2Fchip-seq-chunk-db%2FH3K36me3_TDH_other     1     5</span></span>
<span id="cb13-54"><a href="#cb13-54" tabindex="-1"></a><span class="co">#&gt; 15: HVyz%2FRinst2d4985211eda79%2Fnc%2Fextdata%2Fchip-seq-chunk-db%2FH3K36me3_TDH_other     1     6</span></span>
<span id="cb13-55"><a href="#cb13-55" tabindex="-1"></a><span class="co">#&gt; 16: HVyz%2FRinst2d4985211eda79%2Fnc%2Fextdata%2Fchip-seq-chunk-db%2FH3K36me3_TDH_other     1     7</span></span>
<span id="cb13-56"><a href="#cb13-56" tabindex="-1"></a><span class="co">#&gt; 17: HVyz%2FRinst2d4985211eda79%2Fnc%2Fextdata%2Fchip-seq-chunk-db%2FH3K36me3_TDH_other     1     8</span></span>
<span id="cb13-57"><a href="#cb13-57" tabindex="-1"></a><span class="co">#&gt; 18: HVyz%2FRinst2d4985211eda79%2Fnc%2Fextdata%2Fchip-seq-chunk-db%2FH3K36me3_TDH_other     1     9</span></span>
<span id="cb13-58"><a href="#cb13-58" tabindex="-1"></a><span class="co">#&gt; 19: HVyz%2FRinst2d4985211eda79%2Fnc%2Fextdata%2Fchip-seq-chunk-db%2FH3K36me3_TDH_other     1    10</span></span>
<span id="cb13-59"><a href="#cb13-59" tabindex="-1"></a><span class="co">#&gt; 20: HVyz%2FRinst2d4985211eda79%2Fnc%2Fextdata%2Fchip-seq-chunk-db%2FH3K36me3_TDH_other     1    11</span></span>
<span id="cb13-60"><a href="#cb13-60" tabindex="-1"></a><span class="co">#&gt; 21: HVyz%2FRinst2d4985211eda79%2Fnc%2Fextdata%2Fchip-seq-chunk-db%2FH3K36me3_TDH_other     1    12</span></span>
<span id="cb13-61"><a href="#cb13-61" tabindex="-1"></a><span class="co">#&gt; 22: HVyz%2FRinst2d4985211eda79%2Fnc%2Fextdata%2Fchip-seq-chunk-db%2FH3K4me3_TDH_immune     9     0</span></span>
<span id="cb13-62"><a href="#cb13-62" tabindex="-1"></a><span class="co">#&gt; 23:  HVyz%2FRinst2d4985211eda79%2Fnc%2Fextdata%2Fchip-seq-chunk-db%2FH3K4me3_XJ_immune     2     0</span></span>
<span id="cb13-63"><a href="#cb13-63" tabindex="-1"></a><span class="co">#&gt;                                                                                   data chunk  part</span></span>
<span id="cb13-64"><a href="#cb13-64" tabindex="-1"></a><span class="co">#&gt;      rows</span></span>
<span id="cb13-65"><a href="#cb13-65" tabindex="-1"></a><span class="co">#&gt;     &lt;int&gt;</span></span>
<span id="cb13-66"><a href="#cb13-66" tabindex="-1"></a><span class="co">#&gt;  1:  1000</span></span>
<span id="cb13-67"><a href="#cb13-67" tabindex="-1"></a><span class="co">#&gt;  2:  1000</span></span>
<span id="cb13-68"><a href="#cb13-68" tabindex="-1"></a><span class="co">#&gt;  3:  1000</span></span>
<span id="cb13-69"><a href="#cb13-69" tabindex="-1"></a><span class="co">#&gt;  4:  1000</span></span>
<span id="cb13-70"><a href="#cb13-70" tabindex="-1"></a><span class="co">#&gt;  5:  1000</span></span>
<span id="cb13-71"><a href="#cb13-71" tabindex="-1"></a><span class="co">#&gt;  6:  1000</span></span>
<span id="cb13-72"><a href="#cb13-72" tabindex="-1"></a><span class="co">#&gt;  7:  1000</span></span>
<span id="cb13-73"><a href="#cb13-73" tabindex="-1"></a><span class="co">#&gt;  8:   134</span></span>
<span id="cb13-74"><a href="#cb13-74" tabindex="-1"></a><span class="co">#&gt;  9:  1000</span></span>
<span id="cb13-75"><a href="#cb13-75" tabindex="-1"></a><span class="co">#&gt; 10:  1000</span></span>
<span id="cb13-76"><a href="#cb13-76" tabindex="-1"></a><span class="co">#&gt; 11:  1000</span></span>
<span id="cb13-77"><a href="#cb13-77" tabindex="-1"></a><span class="co">#&gt; 12:  1000</span></span>
<span id="cb13-78"><a href="#cb13-78" tabindex="-1"></a><span class="co">#&gt; 13:  1000</span></span>
<span id="cb13-79"><a href="#cb13-79" tabindex="-1"></a><span class="co">#&gt; 14:  1000</span></span>
<span id="cb13-80"><a href="#cb13-80" tabindex="-1"></a><span class="co">#&gt; 15:  1000</span></span>
<span id="cb13-81"><a href="#cb13-81" tabindex="-1"></a><span class="co">#&gt; 16:  1000</span></span>
<span id="cb13-82"><a href="#cb13-82" tabindex="-1"></a><span class="co">#&gt; 17:  1000</span></span>
<span id="cb13-83"><a href="#cb13-83" tabindex="-1"></a><span class="co">#&gt; 18:  1000</span></span>
<span id="cb13-84"><a href="#cb13-84" tabindex="-1"></a><span class="co">#&gt; 19:  1000</span></span>
<span id="cb13-85"><a href="#cb13-85" tabindex="-1"></a><span class="co">#&gt; 20:  1000</span></span>
<span id="cb13-86"><a href="#cb13-86" tabindex="-1"></a><span class="co">#&gt; 21:   109</span></span>
<span id="cb13-87"><a href="#cb13-87" tabindex="-1"></a><span class="co">#&gt; 22:   886</span></span>
<span id="cb13-88"><a href="#cb13-88" tabindex="-1"></a><span class="co">#&gt; 23:   172</span></span>
<span id="cb13-89"><a href="#cb13-89" tabindex="-1"></a><span class="co">#&gt;      rows</span></span></code></pre></div>
<p>The output above indicates that we have successfully read the data
back into R.</p>
</div>
<div id="example-4-pattern-with-two-more-capture-groups" class="section level2">
<h2>Example 4: pattern with two more capture groups</h2>
<p>In the code below, we read the same data files, with a more complex
pattern that has two additional capture groups (name and id).</p>
<div class="sourceCode" id="cb14"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb14-1"><a href="#cb14-1" tabindex="-1"></a>(count.dt <span class="ot">&lt;-</span> nc<span class="sc">::</span><span class="fu">capture_first_glob</span>(</span>
<span id="cb14-2"><a href="#cb14-2" tabindex="-1"></a>  glob,</span>
<span id="cb14-3"><a href="#cb14-3" tabindex="-1"></a>  data.chunk.pattern,</span>
<span id="cb14-4"><a href="#cb14-4" tabindex="-1"></a>  <span class="st">&quot;/counts/&quot;</span>, </span>
<span id="cb14-5"><a href="#cb14-5" tabindex="-1"></a>  <span class="at">name=</span><span class="fu">list</span>(<span class="st">&quot;McGill&quot;</span>, <span class="at">id=</span><span class="st">&quot;[0-9]+&quot;</span>, as.integer),</span>
<span id="cb14-6"><a href="#cb14-6" tabindex="-1"></a>  <span class="at">READ=</span>read.bedGraph))</span>
<span id="cb14-7"><a href="#cb14-7" tabindex="-1"></a><span class="co">#&gt;                                                                            data chunk       name</span></span>
<span id="cb14-8"><a href="#cb14-8" tabindex="-1"></a><span class="co">#&gt;                                                                          &lt;char&gt; &lt;int&gt;     &lt;char&gt;</span></span>
<span id="cb14-9"><a href="#cb14-9" tabindex="-1"></a><span class="co">#&gt;     1: HVyz/Rinst2d4985211eda79/nc/extdata/chip-seq-chunk-db/H3K36me3_AM_immune     9 McGill0101</span></span>
<span id="cb14-10"><a href="#cb14-10" tabindex="-1"></a><span class="co">#&gt;     2: HVyz/Rinst2d4985211eda79/nc/extdata/chip-seq-chunk-db/H3K36me3_AM_immune     9 McGill0101</span></span>
<span id="cb14-11"><a href="#cb14-11" tabindex="-1"></a><span class="co">#&gt;     3: HVyz/Rinst2d4985211eda79/nc/extdata/chip-seq-chunk-db/H3K36me3_AM_immune     9 McGill0101</span></span>
<span id="cb14-12"><a href="#cb14-12" tabindex="-1"></a><span class="co">#&gt;     4: HVyz/Rinst2d4985211eda79/nc/extdata/chip-seq-chunk-db/H3K36me3_AM_immune     9 McGill0101</span></span>
<span id="cb14-13"><a href="#cb14-13" tabindex="-1"></a><span class="co">#&gt;     5: HVyz/Rinst2d4985211eda79/nc/extdata/chip-seq-chunk-db/H3K36me3_AM_immune     9 McGill0101</span></span>
<span id="cb14-14"><a href="#cb14-14" tabindex="-1"></a><span class="co">#&gt;    ---                                                                                          </span></span>
<span id="cb14-15"><a href="#cb14-15" tabindex="-1"></a><span class="co">#&gt; 20297:  HVyz/Rinst2d4985211eda79/nc/extdata/chip-seq-chunk-db/H3K4me3_XJ_immune     2 McGill0024</span></span>
<span id="cb14-16"><a href="#cb14-16" tabindex="-1"></a><span class="co">#&gt; 20298:  HVyz/Rinst2d4985211eda79/nc/extdata/chip-seq-chunk-db/H3K4me3_XJ_immune     2 McGill0024</span></span>
<span id="cb14-17"><a href="#cb14-17" tabindex="-1"></a><span class="co">#&gt; 20299:  HVyz/Rinst2d4985211eda79/nc/extdata/chip-seq-chunk-db/H3K4me3_XJ_immune     2 McGill0024</span></span>
<span id="cb14-18"><a href="#cb14-18" tabindex="-1"></a><span class="co">#&gt; 20300:  HVyz/Rinst2d4985211eda79/nc/extdata/chip-seq-chunk-db/H3K4me3_XJ_immune     2 McGill0024</span></span>
<span id="cb14-19"><a href="#cb14-19" tabindex="-1"></a><span class="co">#&gt; 20301:  HVyz/Rinst2d4985211eda79/nc/extdata/chip-seq-chunk-db/H3K4me3_XJ_immune     2 McGill0024</span></span>
<span id="cb14-20"><a href="#cb14-20" tabindex="-1"></a><span class="co">#&gt;           id  chrom     start       end count</span></span>
<span id="cb14-21"><a href="#cb14-21" tabindex="-1"></a><span class="co">#&gt;        &lt;int&gt; &lt;char&gt;     &lt;int&gt;     &lt;int&gt; &lt;int&gt;</span></span>
<span id="cb14-22"><a href="#cb14-22" tabindex="-1"></a><span class="co">#&gt;     1:   101  chr10 111456281 111456338     2</span></span>
<span id="cb14-23"><a href="#cb14-23" tabindex="-1"></a><span class="co">#&gt;     2:   101  chr10 111456338 111456381     1</span></span>
<span id="cb14-24"><a href="#cb14-24" tabindex="-1"></a><span class="co">#&gt;     3:   101  chr10 111456381 111459312     0</span></span>
<span id="cb14-25"><a href="#cb14-25" tabindex="-1"></a><span class="co">#&gt;     4:   101  chr10 111459312 111459316     5</span></span>
<span id="cb14-26"><a href="#cb14-26" tabindex="-1"></a><span class="co">#&gt;     5:   101  chr10 111459316 111459409    10</span></span>
<span id="cb14-27"><a href="#cb14-27" tabindex="-1"></a><span class="co">#&gt;    ---                                       </span></span>
<span id="cb14-28"><a href="#cb14-28" tabindex="-1"></a><span class="co">#&gt; 20297:    24  chr22  20689768  20689770     0</span></span>
<span id="cb14-29"><a href="#cb14-29" tabindex="-1"></a><span class="co">#&gt; 20298:    24  chr22  20689770  20689870     1</span></span>
<span id="cb14-30"><a href="#cb14-30" tabindex="-1"></a><span class="co">#&gt; 20299:    24  chr22  20689870  20689995     0</span></span>
<span id="cb14-31"><a href="#cb14-31" tabindex="-1"></a><span class="co">#&gt; 20300:    24  chr22  20689995  20690080     1</span></span>
<span id="cb14-32"><a href="#cb14-32" tabindex="-1"></a><span class="co">#&gt; 20301:    24  chr22  20690080  20691400     0</span></span>
<span id="cb14-33"><a href="#cb14-33" tabindex="-1"></a>count.dt[, .(<span class="at">count=</span>.N), by<span class="ot">=</span>.(data, chunk, name, id, chrom)]</span>
<span id="cb14-34"><a href="#cb14-34" tabindex="-1"></a><span class="co">#&gt;                                                                        data chunk       name    id</span></span>
<span id="cb14-35"><a href="#cb14-35" tabindex="-1"></a><span class="co">#&gt;                                                                      &lt;char&gt; &lt;int&gt;     &lt;char&gt; &lt;int&gt;</span></span>
<span id="cb14-36"><a href="#cb14-36" tabindex="-1"></a><span class="co">#&gt; 1: HVyz/Rinst2d4985211eda79/nc/extdata/chip-seq-chunk-db/H3K36me3_AM_immune     9 McGill0101   101</span></span>
<span id="cb14-37"><a href="#cb14-37" tabindex="-1"></a><span class="co">#&gt; 2: HVyz/Rinst2d4985211eda79/nc/extdata/chip-seq-chunk-db/H3K36me3_TDH_other     1 McGill0019    19</span></span>
<span id="cb14-38"><a href="#cb14-38" tabindex="-1"></a><span class="co">#&gt; 3: HVyz/Rinst2d4985211eda79/nc/extdata/chip-seq-chunk-db/H3K4me3_TDH_immune     9 McGill0024    24</span></span>
<span id="cb14-39"><a href="#cb14-39" tabindex="-1"></a><span class="co">#&gt; 4:  HVyz/Rinst2d4985211eda79/nc/extdata/chip-seq-chunk-db/H3K4me3_XJ_immune     2 McGill0024    24</span></span>
<span id="cb14-40"><a href="#cb14-40" tabindex="-1"></a><span class="co">#&gt;     chrom count</span></span>
<span id="cb14-41"><a href="#cb14-41" tabindex="-1"></a><span class="co">#&gt;    &lt;char&gt; &lt;int&gt;</span></span>
<span id="cb14-42"><a href="#cb14-42" tabindex="-1"></a><span class="co">#&gt; 1:  chr10  7134</span></span>
<span id="cb14-43"><a href="#cb14-43" tabindex="-1"></a><span class="co">#&gt; 2:  chr21 12109</span></span>
<span id="cb14-44"><a href="#cb14-44" tabindex="-1"></a><span class="co">#&gt; 3:   chr1   886</span></span>
<span id="cb14-45"><a href="#cb14-45" tabindex="-1"></a><span class="co">#&gt; 4:  chr22   172</span></span></code></pre></div>
<p>The output above indicates that we have successfully read the data
into R, with two additional columns (name and id). These data can be
visualized using the code below,</p>
<div class="sourceCode" id="cb15"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb15-1"><a href="#cb15-1" tabindex="-1"></a><span class="cf">if</span>(<span class="fu">require</span>(ggplot2)){</span>
<span id="cb15-2"><a href="#cb15-2" tabindex="-1"></a>  <span class="fu">ggplot</span>()<span class="sc">+</span></span>
<span id="cb15-3"><a href="#cb15-3" tabindex="-1"></a>    <span class="fu">facet_wrap</span>(<span class="sc">~</span>data<span class="sc">+</span>chunk<span class="sc">+</span>name<span class="sc">+</span>chrom, <span class="at">labeller=</span>label_both, <span class="at">scales=</span><span class="st">&quot;free&quot;</span>)<span class="sc">+</span></span>
<span id="cb15-4"><a href="#cb15-4" tabindex="-1"></a>    <span class="fu">geom_step</span>(<span class="fu">aes</span>(</span>
<span id="cb15-5"><a href="#cb15-5" tabindex="-1"></a>      start<span class="sc">/</span><span class="fl">1e3</span>, count),</span>
<span id="cb15-6"><a href="#cb15-6" tabindex="-1"></a>      <span class="at">data=</span>count.dt)</span>
<span id="cb15-7"><a href="#cb15-7" tabindex="-1"></a>}</span></code></pre></div>
<p><img src="data:image/png;base64,iVBORw0KGgoAAAANSUhEUgAAASAAAAEgCAMAAAAjXV6yAAAAFVBMVEUAAAAaGhozMzNNTU3Z2dnr6+v////Mrj8CAAAACXBIWXMAAA7DAAAOwwHHb6hkAAANeUlEQVR4nO2di4Krqg5AZWP9/0++bSUhIUAAkXrPJNOxFlFWFw8dHdvtsKjG9muAp4cJUsIEKWGClDBBSpggJXoE7cviAQwmqBHBBCkIJkhBMEEKgglSEEyQgtAjyDn3WdWxDTn5wsllLrtcpcsJWkQxJOjc6mU0PaqC1lCMCNqB7qzE7xRevX/dJ75TTHOQlaR8XoVWwFMh8f1cYXCrKC4Jio/8bOnBn3N5QmKToHspBgSJYnZ4PiuPvNppUTQ3R8PsZEW3t3Sx2ymgwM5BOpJE+Q4bMAWNLd8llIAGbd+R7YZtVQWtoRgRxMo+Z7BQLC+TCdBk497DNkgtnjM1hjUUI4IcK5W0aFoezpIZSsrQODCp4IqgVRQjgqCqsN2GCox1E/cfQOV2ngvbdai4sCHoD+dMmYEVdyvFgKB9WjgleQWDQgEFmqBC8m8FafEABhPUiGCCFAQTpCCYIAXBBCkIJkhBMEEKgglSEEyQgmCCFAQTpCCYIAXhBkHyj2SSgudgmuhGGaoIxT/jswjLBTXh/WcFsctOO5kHJDzn20Y3wKAidDbiyYLSk7v507/NdCOCNIQ2gnsExZPuycPRLLcKakB4uKCmBv5fFlRs3z141wSpCM8cg0iOZrqxSnqyIHLZiV2CQsDbj4N0hF8KmhUPYDBBjQgmSEEwQQqCCVIQTJCCYIIUBBOkIJggBcEEKQgmSEEwQQqCCVIQTJCCYIIUhLsEnafo8QQM3OP1/aH/1U7mM3Q3M+RIUoQlguK9Ak5e4cD5DN3NDDmSFGFUEK2C814QXglwDwD+A/yOCTuc7otk6WnGNQwZkomC8md9YzUBCz/VGQVh3vEWdJUhSzJXEKNkZQS42IJjdU4WNM7ASp/exaB1Ojof27eLJPkuNq8FDTNwQUk/nyIomofaI8tdrNE7BV1gSEu/RRC7vYg2b3KrVqghrNJMw74kaJghRzJVENyQBXBkCEjSHGJOPA66zPDL46Br8QAGE9SIYIIUBBOkIJggBcEEKQgmSEEwQQqCCVIQTJCCYIIUBBOkIJggBcEEKQjTBLnq0iRH7vxCjq6PoQOhcJolg/BDQZV17heUP1GXQZghiH0+UXIZppgjnteq0DUzdCJABTU04imC4rlO8XCZHPg8UVA3wsIWlJzf3DmgzHGDoBEEE1RFYMkLBDmV7tYu1o3gyIq3C6oOALkcO3kLcwT1Ijiy3gpB/K4jekMtNG3HgR2kTRPUh4AXYNcImh8PYDBBjQgmSEEwQQrC0wX5z6/3JihPFwSdT48T5H9WeztjCI4eJ8iTx68EfavHe1JLyxGKgoIcCvd6vT6/3xif2RryRAjvsRX1x4wvT6sIOmtvFK6/xAzDt3qwkv69466i+rcR4I4gqBuuv8Qiw0NbEK+9KbFVS8wwHKta0FbMXBEk+v/lMWjbts/kOxBtLWNQwOippPiG3qXJt5d/v9sFQaO1l+cgEbhqMTAGhQ3D7KmIFFMTlF021sUGOzcXtOXrOIXoG4Nw03H2gOkW9Ml3EJHkBqt4EY7XXpegDZ82GaR1Z/jkcdCLhni1vcJm4VnOvGAKnfyFWb6p782cByL6W83BBaTtnNnOseQsazvLCqCf2SMUjpnzcWY/zhGKjUFwLFaspPB7Pp0tpCPOzYg03vGrR9LJYT6uAVW/IVNox9gqsLwmTomNDSjXzfs0DESTIB8Qee39E3T/xOYvpggM0YIKb0tus22ZXNgmKFt7ci9yX8qJcTTv5mt7uOrer7pQG26TPchyQRmGUvxIEK+83whqO078jaA/HyZICROkRMMgnZl6OMSV8zQF0w++lKZAHi+30xDpZtm0soxdDBALJwiCqRfzNEWmyynmOfhZ3jZD8Zx+9lFdgKWkGTsEFSsnrXN21CTyeJHHy+0Q9szxe4kvqTA6rSxjFwPSjD2Cim0nbDQK+s7TFJ6e5vFHmmdEkIdcnrwEwsoytqKvlzsoiOqgbzvtRCFd5PEiD235nYKgg+YExQEvI8hj2jRB2EnJGxOCCulyyrfTLQhhjuzqtdYVF8Y3NUcQjhdPEHQOZ0QHlRCXiSlbKDJeEgSaSAcR+yZfSJdTlgffIm2oqiNYJTYXtgVfeGS0FsodPg6aPz23TlJaBaW7VLLB+4+D/nyYICVMkBImSAkTpIQJUsIEKWGClDBBSpggJZ7+b8A/YzBBjQgmSEEwQQqCCVIQTJCCYIIUBBOkIHQJyt4J6+SLzE3F6e3rbXQy8JbduynGBOVudx9AU6MuaAnFiCByr3X8kBB45eBmaxc/RsTFDxrZ2aeLxHvC+WeO4B3c5UraF1FcFMTuSs/Mlh78OZfnm1hmYILupBgRJIvZ4fmsPPJqj0Xx3BwNs/PPB68JWkUxJCh8aDUp7GyaTpTgSMt3CSWgQdt3ZLvnenVBayhGBLGyzxksFMvLZAI02bj3sA1Si2oLWkQxIsixUkmLpuXhLJmhpAyNA2MFVwWtoYAC+4+DCAwUGpPJ/gOo3M5zYbsOFRc2BP3hnKkKWkIxJGheuHryEoY6hQlSKH4sSIkHMGQEbSLFBBEd+ZuBTJB6v6sJao9lcE8VZF0sg2BdTEEwQQoCFaQZWgb3WEE2BkkE24spCCZIQbihi8k/AUlK7RNcJd0ow0QE+QE7Nwsq/g2do7tHUBeC1HFFELuospN5IMIzmm10AwyzESYLSk9d5k9uNtONCNIQAKMRQY5B1+D2DBr5boLbu1gDQvJCQZi6F2sR1DRE3iuoZYy+T1CxfffwXROkIDT5qfw1f0FQdQAgOZrpxiqpjtDmp/y32KXdPL2oknzMPDT+u4+DVATX18vtnLSCYIIUhLlj0Gy6hwnSTgktg3uqoMonmpqgY8pebDbdwwSJFBNkglQE24spCLYXUxDsnLSCYIIUBBOkIJggBcEEKQh3CQqnXuAV3MH0/aFfl0v+fztD90OGpYLif8K77Om+FYJ6Ga4KSqsgXo2iPC7+e/eOCTuc3Iu3jqSn+B7AcFlQ/qxvrCZg4aeDEzgCPSLoToY5ghhlLC/CYdt1sTo5nMvAdQm6iWGOIJecIo/N1EWSSvPG3yst6B6GKYLilqH2yHIXa7QMFyowvc7wAIZZgtjNM7R5kxuRQvvGKmVr4ObGBd3CMEUQ3G4ERZEhIElziMmOQbDeLnWxWxiuCro3HsBgghoRTJCCYIIUBBOkIJggBcEEKQgmSEEwQQqCCVIQTJCCYIIUBBOkIJggBcEEKQgzBTktA81RzTzG0FE+vyJWQ/irgsijjjBJEPuEouRSVTlHMboZOsvnJ+2XCIrng8XDZXLMbkH95bPp7YKSc8A7Z5Q5ZgsaKZ9MTZAsX2k/dwhyKuB9Xay7fNXPyjEol2PlGJTJofu5Zy+GD3rTMdA4zjxZUF/54VM3a3eOTRY0OR7AYIIaEUyQgvBwQeK7K5cjVAT9Dg7p5LefLkcoC/oh3B4Rjvi9sQ8T9Eu4KIh+7fnjBLHvgv/EC6JxrnsFnAOG8G3HOQayVuZl+6vSQl0Q+Srmd/x7h8jTM8KPBP1a6RJDZ1RviMuuUFrwrTnevIXm7bVtlQZxvQVJhqux9SqqtCCfNu9saT2FdYdk6O0p6at3hR6xVitZ43sswrV0saWCJnSxcGd3zxo1tlLtYRf7RLnHTOhiR7Gbj7Wg7ftqSgtq6GLaTcCXo6WbdwR+bkIPdC2r1sXUz/SqFNu61qS9WPJp4nME1Y6DPo10216h1EKPqXSi9zoyTXSxGkNrF9uQEUaD48X62IVB+jgyu9jw+RVndaRtCGcLLQQr8FxpYwsyqxQYtAhgJ2eOpaPda39qiOa9kWAvJkUDQ9tmxNZSRU0I146D3sxtaY3Z0rQqQ3U4qi3sW1FpalrzvldQleEZgtjpjsbCZgsqMDxFUP0QZI2gLMNDBP35MEFKmCAlugWFAYGc0ec745gGUxxDZLZcGq6aKTZb+EHLSpexArpWDKX2C8o88OxjTKMnS2AvFNNFmrhGkO65csXGso6kLMomCm1cMZTaK+g8KPFsPv5gGr5vTEIeku2Av7VovuyfF7FY2Cp/L7llsH3GxxaKFWlKmIyMQT56xw1n0jz0FgKVZAsVzPOJKxmk2LgFIcj7ZBl5n/gjFvIVpwjy2DiRDgYhngZtG1NlNo9DBHu/2fPQeOoMNskGjYMXRxQenupJFooVp7UgKIq1IJ4GZXMBNFsujRXACk0qWL4jny7z3lM2uVCuSFLHBCWNkBUfe3Icg+Jb9t7LbDKtIAiBvZzCMs+X4Za4ILJQrJgUPtSCfOZR2It51nRj+lFLS/Z81E++8LCddKeJHFBjnStCeYPHQemBC/igadhnuo+DyPJYatzhicJxIJM5PEbnitDj7EhaCROkhAlSwgQpYYKUMEFKPF4Qvei2yaeF5T8i5DUpvKQYLgHyp9VEP4+coC3+z8qWPq0m+mF8L2nCJdEDLh+bIIyNPuAXetcRrvZvydMarmfEBlPZUkzQJzZ8z9C1TFASpFNlBP31MSg4iG0jtBS68E8LCmMzGaSPVNAfPw6C/3o6zt09jEbnMsghn+7nWlLK/3GYICVMkBImSAkTpIQJUsIEKWGClDBBSpggJUyQEv8DGPD9nbhqSYIAAAAASUVORK5CYII=" /><!-- --></p>
<p>The plot above includes panel/facet titles which come from the
variables which were stored in the file names.</p>
</div>
<div id="example-5-parsing-non-csv-data" class="section level2">
<h2>Example 5: parsing non-CSV data</h2>
<p>The following example demonstrates how non-CSV data may be parsed,
using a custom <code>READ</code> function. Consider the vignette data
files,</p>
<div class="sourceCode" id="cb16"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb16-1"><a href="#cb16-1" tabindex="-1"></a>vignettes <span class="ot">&lt;-</span> <span class="fu">system.file</span>(<span class="st">&quot;extdata/vignettes&quot;</span>, <span class="at">package=</span><span class="st">&quot;nc&quot;</span>, <span class="at">mustWork=</span><span class="cn">TRUE</span>)</span>
<span id="cb16-2"><a href="#cb16-2" tabindex="-1"></a>(vglob <span class="ot">&lt;-</span> <span class="fu">paste0</span>(vignettes, <span class="st">&quot;/*.Rmd&quot;</span>))</span>
<span id="cb16-3"><a href="#cb16-3" tabindex="-1"></a><span class="co">#&gt; [1] &quot;/tmp/RtmppDHVyz/Rinst2d4985211eda79/nc/extdata/vignettes/*.Rmd&quot;</span></span>
<span id="cb16-4"><a href="#cb16-4" tabindex="-1"></a>(vfiles <span class="ot">&lt;-</span> <span class="fu">Sys.glob</span>(vglob))</span>
<span id="cb16-5"><a href="#cb16-5" tabindex="-1"></a><span class="co">#&gt; [1] &quot;/tmp/RtmppDHVyz/Rinst2d4985211eda79/nc/extdata/vignettes/v0-overview.Rmd&quot;     </span></span>
<span id="cb16-6"><a href="#cb16-6" tabindex="-1"></a><span class="co">#&gt; [2] &quot;/tmp/RtmppDHVyz/Rinst2d4985211eda79/nc/extdata/vignettes/v1-capture-first.Rmd&quot;</span></span>
<span id="cb16-7"><a href="#cb16-7" tabindex="-1"></a><span class="co">#&gt; [3] &quot;/tmp/RtmppDHVyz/Rinst2d4985211eda79/nc/extdata/vignettes/v2-capture-all.Rmd&quot;  </span></span>
<span id="cb16-8"><a href="#cb16-8" tabindex="-1"></a><span class="co">#&gt; [4] &quot;/tmp/RtmppDHVyz/Rinst2d4985211eda79/nc/extdata/vignettes/v3-capture-melt.Rmd&quot; </span></span>
<span id="cb16-9"><a href="#cb16-9" tabindex="-1"></a><span class="co">#&gt; [5] &quot;/tmp/RtmppDHVyz/Rinst2d4985211eda79/nc/extdata/vignettes/v4-comparisons.Rmd&quot;  </span></span>
<span id="cb16-10"><a href="#cb16-10" tabindex="-1"></a><span class="co">#&gt; [6] &quot;/tmp/RtmppDHVyz/Rinst2d4985211eda79/nc/extdata/vignettes/v5-helpers.Rmd&quot;      </span></span>
<span id="cb16-11"><a href="#cb16-11" tabindex="-1"></a><span class="co">#&gt; [7] &quot;/tmp/RtmppDHVyz/Rinst2d4985211eda79/nc/extdata/vignettes/v6-engines.Rmd&quot;</span></span></code></pre></div>
<p>The output above includes the glob and the files it matches. Below we
define a function for parsing one of those files,</p>
<div class="sourceCode" id="cb17"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb17-1"><a href="#cb17-1" tabindex="-1"></a>non.greedy.lines <span class="ot">&lt;-</span> <span class="fu">list</span>(</span>
<span id="cb17-2"><a href="#cb17-2" tabindex="-1"></a>  <span class="fu">list</span>(<span class="st">&quot;.*</span><span class="sc">\n</span><span class="st">&quot;</span>), <span class="st">&quot;*?&quot;</span>)</span>
<span id="cb17-3"><a href="#cb17-3" tabindex="-1"></a>optional.name <span class="ot">&lt;-</span> <span class="fu">list</span>(</span>
<span id="cb17-4"><a href="#cb17-4" tabindex="-1"></a>  <span class="fu">list</span>(<span class="st">&quot; &quot;</span>, <span class="at">chunk_name=</span><span class="st">&quot;[^,}]+&quot;</span>), <span class="st">&quot;?&quot;</span>)</span>
<span id="cb17-5"><a href="#cb17-5" tabindex="-1"></a>chunk.pattern <span class="ot">&lt;-</span> <span class="fu">list</span>(</span>
<span id="cb17-6"><a href="#cb17-6" tabindex="-1"></a>  <span class="at">before=</span>non.greedy.lines,</span>
<span id="cb17-7"><a href="#cb17-7" tabindex="-1"></a>  <span class="st">&quot;```</span><span class="sc">\\</span><span class="st">{r&quot;</span>,</span>
<span id="cb17-8"><a href="#cb17-8" tabindex="-1"></a>  optional.name,</span>
<span id="cb17-9"><a href="#cb17-9" tabindex="-1"></a>  <span class="at">parameters=</span><span class="st">&quot;.*&quot;</span>,</span>
<span id="cb17-10"><a href="#cb17-10" tabindex="-1"></a>  <span class="st">&quot;</span><span class="sc">\\</span><span class="st">}</span><span class="sc">\n</span><span class="st">&quot;</span>,</span>
<span id="cb17-11"><a href="#cb17-11" tabindex="-1"></a>  <span class="at">code=</span>non.greedy.lines,</span>
<span id="cb17-12"><a href="#cb17-12" tabindex="-1"></a>  <span class="st">&quot;```&quot;</span>)</span>
<span id="cb17-13"><a href="#cb17-13" tabindex="-1"></a>READ.vignette <span class="ot">&lt;-</span> <span class="cf">function</span>(f)nc<span class="sc">::</span><span class="fu">capture_all_str</span>(f, chunk.pattern)</span>
<span id="cb17-14"><a href="#cb17-14" tabindex="-1"></a><span class="fu">str</span>(<span class="fu">READ.vignette</span>(vfiles[<span class="dv">1</span>]))</span></code></pre></div>
<pre><code>#&gt; Classes &#39;data.table&#39; and &#39;data.frame&#39;:   7 obs. of  4 variables:
#&gt;  $ before    : chr  &quot;&lt;!--\n%\\VignetteEngine{knitr::knitr}\n%\\VignetteIndexEntry{vignette 0: Overview}\n--&gt;\n\n# Overview of nc functionality\n\n&quot; &quot;\n\nHere is an index of topics which are explained in the different\nvignettes, along with an overview of funct&quot;| __truncated__ &quot;\n\nA variant is doing the same thing, but with input\nsubjects coming from a data table/frame with character columns.\n\n&quot; &quot;\n\n## Capture all matches in a single subject\n  \n[Capture all](v2-capture-all.html) is for the situation whe&quot;| __truncated__ ...
#&gt;  $ chunk_name: chr  &quot;setup&quot; &quot;&quot; &quot;&quot; &quot;&quot; ...
#&gt;  $ parameters: chr  &quot;, include = FALSE&quot; &quot;&quot; &quot;&quot; &quot;&quot; ...
#&gt;  $ code      : chr  &quot;knitr::opts_chunk$set(\n  collapse = TRUE,\n  comment = \&quot;#&gt;\&quot;\n)\n&quot; &quot;subject.vec &lt;- c(\n  \&quot;chr10:213054000-213,055,000\&quot;,\n  \&quot;chrM:111000\&quot;,\n  \&quot;chr1:110-111 chr2:220-222\&quot;)\nnc&quot;| __truncated__ &quot;subject.dt &lt;- data.table::data.table(\n  JobID = c(\&quot;13937810_25\&quot;, \&quot;14022192_1\&quot;),\n  Elapsed = c(\&quot;07:04:42\&quot;| __truncated__ &quot;nc::capture_all_str(\n  subject.vec, chrom=\&quot;chr.*?\&quot;, \&quot;:\&quot;, chromStart=\&quot;[0-9,]+\&quot;, as.integer)\n&quot; ...
#&gt;  - attr(*, &quot;.internal.selfref&quot;)=&lt;externalptr&gt;</code></pre>
<p>The output above shows a data table with 7 rows, one for each code
chunk defined in the vignette data file. We read all of the vignette
files using the code below.</p>
<div class="sourceCode" id="cb19"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb19-1"><a href="#cb19-1" tabindex="-1"></a>chunk.dt <span class="ot">&lt;-</span> nc<span class="sc">::</span><span class="fu">capture_first_glob</span>(</span>
<span id="cb19-2"><a href="#cb19-2" tabindex="-1"></a>  vglob,</span>
<span id="cb19-3"><a href="#cb19-3" tabindex="-1"></a>  <span class="st">&quot;/v&quot;</span>,</span>
<span id="cb19-4"><a href="#cb19-4" tabindex="-1"></a>  <span class="at">vignette_number=</span><span class="st">&quot;[0-9]&quot;</span>, as.integer,</span>
<span id="cb19-5"><a href="#cb19-5" tabindex="-1"></a>  <span class="st">&quot;-&quot;</span>,</span>
<span id="cb19-6"><a href="#cb19-6" tabindex="-1"></a>  <span class="at">vignette_name=</span><span class="st">&quot;.*?&quot;</span>,</span>
<span id="cb19-7"><a href="#cb19-7" tabindex="-1"></a>  <span class="st">&quot;.Rmd&quot;</span>,</span>
<span id="cb19-8"><a href="#cb19-8" tabindex="-1"></a>  <span class="at">READ=</span>READ.vignette</span>
<span id="cb19-9"><a href="#cb19-9" tabindex="-1"></a>)[</span>
<span id="cb19-10"><a href="#cb19-10" tabindex="-1"></a>, chunk_number <span class="sc">:=</span> <span class="fu">seq_along</span>(chunk_name), by<span class="ot">=</span>vignette_number</span>
<span id="cb19-11"><a href="#cb19-11" tabindex="-1"></a>]</span>
<span id="cb19-12"><a href="#cb19-12" tabindex="-1"></a>chunk.dt[, .(</span>
<span id="cb19-13"><a href="#cb19-13" tabindex="-1"></a>  vignette_number, vignette_name, chunk_number, chunk_name, </span>
<span id="cb19-14"><a href="#cb19-14" tabindex="-1"></a>  <span class="at">lines=</span><span class="fu">nchar</span>(code))]</span>
<span id="cb19-15"><a href="#cb19-15" tabindex="-1"></a><span class="co">#&gt;      vignette_number vignette_name chunk_number chunk_name lines</span></span>
<span id="cb19-16"><a href="#cb19-16" tabindex="-1"></a><span class="co">#&gt;                &lt;int&gt;        &lt;char&gt;        &lt;int&gt;     &lt;char&gt; &lt;int&gt;</span></span>
<span id="cb19-17"><a href="#cb19-17" tabindex="-1"></a><span class="co">#&gt;   1:               0      overview            1      setup    61</span></span>
<span id="cb19-18"><a href="#cb19-18" tabindex="-1"></a><span class="co">#&gt;   2:               0      overview            2              192</span></span>
<span id="cb19-19"><a href="#cb19-19" tabindex="-1"></a><span class="co">#&gt;   3:               0      overview            3              314</span></span>
<span id="cb19-20"><a href="#cb19-20" tabindex="-1"></a><span class="co">#&gt;   4:               0      overview            4               91</span></span>
<span id="cb19-21"><a href="#cb19-21" tabindex="-1"></a><span class="co">#&gt;   5:               0      overview            5              242</span></span>
<span id="cb19-22"><a href="#cb19-22" tabindex="-1"></a><span class="co">#&gt;  ---                                                            </span></span>
<span id="cb19-23"><a href="#cb19-23" tabindex="-1"></a><span class="co">#&gt; 104:               6       engines            1      setup    61</span></span>
<span id="cb19-24"><a href="#cb19-24" tabindex="-1"></a><span class="co">#&gt; 105:               6       engines            2              115</span></span>
<span id="cb19-25"><a href="#cb19-25" tabindex="-1"></a><span class="co">#&gt; 106:               6       engines            3               67</span></span>
<span id="cb19-26"><a href="#cb19-26" tabindex="-1"></a><span class="co">#&gt; 107:               6       engines            4              122</span></span>
<span id="cb19-27"><a href="#cb19-27" tabindex="-1"></a><span class="co">#&gt; 108:               6       engines            5              320</span></span></code></pre></div>
<p>The output above is a data table with one row for each chunk in each
data file. Some columns (<code>vignette_number</code> and
<code>vignette_name</code>) come from the file path, and others come
from the data file contents, including chunk number, name, and line
count. The files also contain code which has been parsed and can be
extracted via the code below, for example:</p>
<div class="sourceCode" id="cb20"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb20-1"><a href="#cb20-1" tabindex="-1"></a><span class="fu">cat</span>(chunk.dt<span class="sc">$</span>code[<span class="dv">2</span>])</span>
<span id="cb20-2"><a href="#cb20-2" tabindex="-1"></a><span class="co">#&gt; subject.vec &lt;- c(</span></span>
<span id="cb20-3"><a href="#cb20-3" tabindex="-1"></a><span class="co">#&gt;   &quot;chr10:213054000-213,055,000&quot;,</span></span>
<span id="cb20-4"><a href="#cb20-4" tabindex="-1"></a><span class="co">#&gt;   &quot;chrM:111000&quot;,</span></span>
<span id="cb20-5"><a href="#cb20-5" tabindex="-1"></a><span class="co">#&gt;   &quot;chr1:110-111 chr2:220-222&quot;)</span></span>
<span id="cb20-6"><a href="#cb20-6" tabindex="-1"></a><span class="co">#&gt; nc::capture_first_vec(</span></span>
<span id="cb20-7"><a href="#cb20-7" tabindex="-1"></a><span class="co">#&gt;   subject.vec, chrom=&quot;chr.*?&quot;, &quot;:&quot;, chromStart=&quot;[0-9,]+&quot;, as.integer)</span></span></code></pre></div>
</div>
<div id="conclusion" class="section level2">
<h2>Conclusion</h2>
<p>In this vignette we have seen how to read regularly named data files
into R, by providing a glob and a regular expression to
<code>nc::capture_first_glob</code>.</p>
<div class="sourceCode" id="cb21"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb21-1"><a href="#cb21-1" tabindex="-1"></a><span class="fu">options</span>(old.opt)</span></code></pre></div>
</div>



<!-- code folding -->


<!-- dynamically load mathjax for compatibility with self-contained -->
<script>
  (function () {
    var script = document.createElement("script");
    script.type = "text/javascript";
    script.src  = "https://mathjax.rstudio.com/latest/MathJax.js?config=TeX-AMS-MML_HTMLorMML";
    document.getElementsByTagName("head")[0].appendChild(script);
  })();
</script>

</body>
</html>
